---
title: "01_annotation"
author: "Tang Li"
date: '2022-11-23'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(GenomeInfoDb)
library(rtracklayer)
library(stringr)
library(venn)
library(ttFunctions)

require(knitr)
opts_knit$set(root.dir = "~/reference/annotation/")
```

### genename
```{r}
# 把mm9\mm10\mm39的基因名对应起来
gtf_mm9 <- import("mm9/Mus_musculus.NCBIM37.67.gtf")
gtf_mm10 <- import("mm10/Mus_musculus.GRCm38.102.gtf")
gtf_mm39 <- import("mm39/Mus_musculus.GRCm39.108.gtf")

table(duplicated(gtf_mm9$gene_id))
gtf_mm9_gene <- as.data.frame(gtf_mm9[!duplicated(gtf_mm9$gene_id),])
gtf_mm10_gene <- as.data.frame(gtf_mm10[gtf_mm10$type=="gene",])
gtf_mm39_gene <- as.data.frame(gtf_mm39[gtf_mm39$type=="gene",])

venn(list(mm9=gtf_mm9_gene$gene_id,
          mm10=gtf_mm10_gene$gene_id,
          mm39=gtf_mm39_gene$gene_id))

venn(list(mm9=gtf_mm9_gene$gene_name,
          mm10=gtf_mm10_gene$gene_name,
          mm39=gtf_mm39_gene$gene_name))

gtf_mm9_gene_pc <- gtf_mm9_gene[gtf_mm9_gene$gene_biotype=="protein_coding",]
gtf_mm10_gene_pc <- gtf_mm10_gene[gtf_mm10_gene$gene_biotype=="protein_coding",]
gtf_mm39_gene_pc <- gtf_mm39_gene[gtf_mm39_gene$gene_biotype=="protein_coding",]


venn(list(mm9_pc=gtf_mm9_gene_pc$gene_id,
          mm10_pc=gtf_mm10_gene_pc$gene_id,
          mm39_pc=gtf_mm39_gene_pc$gene_id))

venn(list(mm9_pc=gtf_mm9_gene_pc$gene_name,
          mm10_pc=gtf_mm10_gene_pc$gene_name,
          mm39_pc=gtf_mm39_gene_pc$gene_name))

gene_id_name <- merge(gtf_mm9_gene[,c(10,13)],gtf_mm10_gene[,c(10,12)],by="gene_id",all=T)

gene_id_name <- merge(gene_id_name,gtf_mm39_gene[,c(10,12)],by="gene_id",all=T)
colnames(gene_id_name) <- c("gene_id","mm9","mm10","mm39")

#gene_id_name$mm9 <- str_to_title(gene_id_name$mm9)
#gene_id_name$mm10 <- str_to_title(gene_id_name$mm10)
#gene_id_name$mm39 <- str_to_title(gene_id_name$mm39)

tt_wt(gene_id_name,"gene_id_name_mm.txt",col.names=T)

gene_id_name$all <- paste(gene_id_name$mm9,gene_id_name$mm10,gene_id_name$mm39,sep = "_")

table(is.na(gene_id_name$all))
table(duplicated(gene_id_name$all))

gene_id_name_dedup <- gene_id_name[!duplicated(gene_id_name$all),]

gene_id_name_final <- gene_id_name_dedup[rowSums(is.na(gene_id_name_dedup))==0,1:4]
gene_id_name_NA1 <- gene_id_name_dedup[rowSums(is.na(gene_id_name_dedup))==1,1:4]
gene_id_name_NA2 <- gene_id_name_dedup[rowSums(is.na(gene_id_name_dedup))==2,1:4]

gene_id_name_dedup <- gene_id_name_dedup[!rowSums(is.na(gene_id_name_dedup))==0,1:4]

for (i in 1:dim(gene_id_name_NA1)[1]){
  if (is.na(gene_id_name_NA1[i,"mm9"])){
  if (gene_id_name_NA1[i,"mm10"] %in% gene_id_name_dedup$mm9){
    gene_id_name_NA1[i,"mm9"] <- gene_id_name_NA1[i,"mm10"]
  } else if (gene_id_name_NA1[i,"mm39"] %in% gene_id_name_dedup$mm9){
    gene_id_name_NA1[i,"mm9"] <- gene_id_name_NA1[i,"mm39"]
  }
} else if (is.na(gene_id_name_NA1[i,"mm39"])){
  if (gene_id_name_NA1[i,"mm10"] %in% gene_id_name_dedup$mm39){
    gene_id_name_NA1[i,"mm39"] <- gene_id_name_NA1[i,"mm10"]
  } else if (gene_id_name_NA1[i,"mm9"] %in% gene_id_name_dedup$mm39){
    gene_id_name_NA1[i,"mm39"] <- gene_id_name_NA1[i,"mm9"]
  }
}
  if (i %% 1000==0){
    print(i)
  }
}

table(rowSums(is.na(gene_id_name_NA1)))

gene_id_name_final <- rbind(gene_id_name_final,gene_id_name_NA1)

for (i in 1:dim(gene_id_name_NA2)[1]){
  if (!is.na(gene_id_name_NA2[i,"mm9"])){
    if (gene_id_name_NA2[i,"mm9"] %in% gene_id_name_NA2$mm10){
      gene_id_name_NA2[i,"mm10"] <- gene_id_name_NA2[i,"mm9"]
    } 
    if (!gene_id_name_NA2[i,"mm9"] %in% gene_id_name_NA2$mm39){
      gene_id_name_NA2[i,"mm39"] <- gene_id_name_NA2[i,"mm9"]
    }
  } else if (!is.na(gene_id_name_NA2[i,"mm39"])){
    if (gene_id_name_NA2[i,"mm39"] %in% gene_id_name_NA2$mm10){
      gene_id_name_NA2[i,"mm10"] <- gene_id_name_NA2[i,"mm39"]
    }
    if (gene_id_name_NA2[i,"mm39"] %in% gene_id_name_NA2$mm9){
    gene_id_name_NA2[i,"mm9"] <- gene_id_name_NA2[i,"mm39"]
    }
  } else if (!is.na(gene_id_name_NA2[i,"mm10"])){
    if (gene_id_name_NA2[i,"mm10"] %in% gene_id_name_NA2$mm39){
      gene_id_name_NA2[i,"mm39"] <- gene_id_name_NA2[i,"mm10"]
    }
    if (gene_id_name_NA2[i,"mm10"] %in% gene_id_name_NA2$mm9){
    gene_id_name_NA2[i,"mm9"] <- gene_id_name_NA2[i,"mm10"]
    }
  }
  if (i %% 1000==0){
    print(i)
  }
}
table(rowSums(is.na(gene_id_name_NA2)))

gene_id_name_final <- rbind(gene_id_name_final,gene_id_name_NA2)

gene_id_name_final$all <- paste(gene_id_name_final$mm9,gene_id_name_final$mm10,gene_id_name_final$mm39,sep = "_")

gene_id_name_dedup2 <- gene_id_name_final[!duplicated(gene_id_name_final$all),]

# table(gene_id_name_dedup2$mm39 %in% gene_id_name_final$mm39)
# gene_id_name_dedup <- gene_id_name_dedup[!(gene_id_name_dedup$mm39 %in% gene_id_name_final$mm39),]
# 
# table(gene_id_name_dedup$mm10 %in% gene_id_name_final$mm10)
# gene_id_name_dedup <- gene_id_name_dedup[!(gene_id_name_dedup$mm10 %in% gene_id_name_final$mm10),]
# 
# table(gene_id_name_dedup$mm9 %in% gene_id_name_final$mm9)
# gene_id_name_dedup <- gene_id_name_dedup[!(gene_id_name_dedup$mm9 %in% gene_id_name_final$mm9),]
# 
# 
# venn(list(gene_id_name_dedup$mm9,gene_id_name_dedup$mm10,gene_id_name_dedup$mm39))
# 
# gene_name_all <- c(gene_id_name_dedup$mm9,gene_id_name_dedup$mm10,gene_id_name_dedup$mm39)
# 
# gene_name_all <- gene_name_all[!duplicated(gene_name_all)]
# 
# table(is.na(gene_name_all))
# 
# gene_name_all <- gene_name_all[!is.na(gene_name_all)]
# 
# j=1
# for (i in gene_name_all){
#   if(j %% 1000==0){
#     print(j)
#   }
#   if ((i %in% gene_id_name_dedup$mm9)&(i %in% gene_id_name_dedup$mm10)&(i %in% gene_id_name_dedup$mm39)){
#     gene_id_name_final <- rbind(gene_id_name_final,c(paste0("mamual_",i),i,i,i))
#   } else if ((i %in% gene_id_name_dedup$mm9)&(i %in% gene_id_name_dedup$mm10)){
#     gene_id_name_final <- rbind(gene_id_name_final,c(paste0("mamual_",i),i,i,NA))
#   } else if ((i %in% gene_id_name_dedup$mm9)&(i %in% gene_id_name_dedup$mm39)){
#     gene_id_name_final <- rbind(gene_id_name_final,c(paste0("mamual_",i),i,NA,i))
#   } else if ((i %in% gene_id_name_dedup$mm10)&(i %in% gene_id_name_dedup$mm39)){
#     gene_id_name_final <- rbind(gene_id_name_final,c(paste0("mamual_",i),NA,i,i))
#   } else if (i %in% gene_id_name_dedup$mm9){
#     gene_id_name_final <- rbind(gene_id_name_final,c(paste0("mamual_",i),i,NA,NA))
#   } else if (i %in% gene_id_name_dedup$mm10){
#     gene_id_name_final <- rbind(gene_id_name_final,c(paste0("mamual_",i),NA,i,NA))
#   } else if (i %in% gene_id_name_dedup$mm39){
#     gene_id_name_final <- rbind(gene_id_name_final,c(paste0("mamual_",i),NA,NA,i))
#   }
#   j <- j+1
# }
# 
# table(duplicated(gene_id_name_final$mm39[!is.na(gene_id_name_final$mm39)]))
# gene_id_name_final$mm39[!is.na(gene_id_name_final$mm39)][duplicated(gene_id_name_final$mm39[!is.na(gene_id_name_final$mm39)])]
# 
# table(duplicated(gene_id_name_final$mm10[!is.na(gene_id_name_final$mm10)]))
# gene_id_name_final$mm10[!is.na(gene_id_name_final$mm10)][duplicated(gene_id_name_final$mm10[!is.na(gene_id_name_final$mm10)])]
# 
# table(duplicated(gene_id_name_final$mm9[!is.na(gene_id_name_final$mm9)]))
# gene_id_name_final$mm9[!is.na(gene_id_name_final$mm9)][duplicated(gene_id_name_final$mm9[!is.na(gene_id_name_final$mm9)])]

tt_wt(gene_id_name_dedup2[,2:4],"gene_id_name_final2_mm.txt",col.names = T)

venn(list(gene_id_name$mm9,gene_id_name_final$mm9))
venn(list(gene_id_name$mm10,gene_id_name_final$mm10))
venn(list(gene_id_name$mm39,gene_id_name_final$mm39))





gene_id_name_dedup[gene_id_name_dedupintersect(intersect(gene_id_name_dedup$mm9,gene_id_name_dedup$mm10),gene_id_name_dedup$mm39),]

gene_id_name_dedup[(gene_id_name_dedup$mm9 %in% gene_id_name_final$mm9),]

gene_id_name_dedup[(gene_id_name_dedup$mm9 %in% gene_id_name_dedup[duplicated(gene_id_name_dedup$mm9),"mm9"]) & rowSums(is.na(gene_id_name_dedup))==0,]

table(duplicated(gene_id_name_dedup[!is.na(gene_id_name_dedup$mm9),]$mm9))
gene_id_name_dedup[duplicated(gene_id_name_dedup[!is.na(gene_id_name_dedup$mm9),]$mm9),]
rowSums(is.na(gene_id_name_dedup))

table(rowSums(is.na(gene_id_name_dedup)))


```

## hg38
```{r}
## read gtf
hg38_gtf_gencode <- import("~/reference/annotation/hg38/gencode.v40.basic.annotation.gtf")
hg38_gtf_ensembl <- import("~/reference/annotation/hg38/Homo_sapiens.GRCh38.108.gtf")

head(hg38_gtf_gencode$gene_id)
head(hg38_gtf_ensembl$gene_id)

## overlap between gencode and ensembl
hg38_gtf_gencode_gene_id <- sapply(strsplit(hg38_gtf_gencode$gene_id,split = ".",fixed=T),"[",1)
hg38_gtf_gencode_gene_id <- hg38_gtf_gencode_gene_id[!duplicated(hg38_gtf_gencode_gene_id)]

hg38_gtf_ensembl_gene_id <- hg38_gtf_ensembl$gene_id
hg38_gtf_ensembl_gene_id <- hg38_gtf_ensembl_gene_id[!duplicated(hg38_gtf_ensembl_gene_id)]

venn(list(hg38_gtf_gencode_gene_id,hg38_gtf_ensembl_gene_id))

head(hg38_gtf_gencode$transcript_id)
head(hg38_gtf_ensembl$transcript_id)


hg38_gtf_gencode_transcript_id <- sapply(strsplit(hg38_gtf_gencode$transcript_id,split = ".",fixed=T),"[",1)
hg38_gtf_gencode_transcript_id <- hg38_gtf_gencode_transcript_id[!duplicated(hg38_gtf_gencode_transcript_id)]

hg38_gtf_ensembl_transcript_id <- hg38_gtf_ensembl$transcript_id
hg38_gtf_ensembl_transcript_id <- hg38_gtf_ensembl_transcript_id[!duplicated(hg38_gtf_ensembl_transcript_id)]

venn(list(hg38_gtf_gencode_transcript_id,hg38_gtf_ensembl_transcript_id))

## chr modified 
table(hg38_gtf_gencode[hg38_gtf_gencode$type=="gene",]@seqnames)
table(hg38_gtf_ensembl[hg38_gtf_ensembl$type=="gene",]@seqnames)
chr_name <- c(seq(1,22),"X","Y","MT")
hg38_gtf_ensembl_chr <- hg38_gtf_ensembl[hg38_gtf_ensembl@seqnames %in% chr_name,]
table(hg38_gtf_ensembl_chr[hg38_gtf_ensembl_chr$type=="gene",]@seqnames)
tail(hg38_gtf_ensembl_chr[hg38_gtf_ensembl_chr$type=="gene",]@seqnames)

hg38_gtf_ensembl_chr@seqnames <- Rle(factor(paste0("chr",hg38_gtf_ensembl_chr@seqnames),levels = paste0("chr",c(seq(1,22),"X","Y","MT"))))
tail(hg38_gtf_ensembl_chr[hg38_gtf_ensembl_chr$type=="gene",]@seqnames)
hg38_gtf_ensembl_chr@seqnames <- Rle(factor(ifelse(hg38_gtf_ensembl_chr@seqnames=="chrMT","chrM",hg38_gtf_ensembl_chr@seqnames),levels = paste0("chr",c(seq(1,22),"X","Y","M"))))
tail(hg38_gtf_ensembl_chr[hg38_gtf_ensembl_chr$type=="gene",]@seqnames)

export(hg38_gtf_ensembl_chr,"~/reference/annotation/hg38/Homo_sapiens.GRCh38.108.chr.gtf","gtf")
```

```{bash}
sed 's/three_prime_utr/3UTR/g' Homo_sapiens.GRCh38.108.chr.gtf | sed 's/five_prime_utr/5UTR/g' > Homo_sapiens.GRCh38.108.chr_UTR.gtf 
```

```{r basic pc_lnc}

hg38_gtf_ensembl_chr <- import("~/reference/annotation/hg38/Homo_sapiens.GRCh38.108.chr_UTR.gtf")

table(hg38_gtf_ensembl_chr$type)

hg38_gtf_ensembl_basic <- hg38_gtf_ensembl_chr[hg38_gtf_ensembl_chr$gene_id %in% hg38_gtf_gencode_gene_id,]
hg38_gtf_ensembl_basic <- hg38_gtf_ensembl_basic[hg38_gtf_ensembl_basic$transcript_id %in% hg38_gtf_gencode_transcript_id,]

export(hg38_gtf_ensembl_basic,"~/reference/annotation/hg38/Homo_sapiens.GRCh38.108.basic.gtf","gtf")

hg38_gtf_ensembl_basic_pc_lnc <- hg38_gtf_ensembl_basic[hg38_gtf_ensembl_basic$gene_biotype %in% c("lncRNA","protein_coding"),]
export(hg38_gtf_ensembl_basic_pc_lnc,"~/reference/annotation/hg38/Homo_sapiens.GRCh38.108.basic.pc_lnc.gtf","gtf")

hg38_gtf_ensembl_basic_pc <- hg38_gtf_ensembl_basic[hg38_gtf_ensembl_basic$gene_biotype %in% c("protein_coding"),]
export(hg38_gtf_ensembl_basic_pc,"~/reference/annotation/hg38/Homo_sapiens.GRCh38.108.basic.pc.gtf","gtf")

```

```{r transcript_id gene_name}
dim(as.data.frame(hg38_gtf_ensembl[hg38_gtf_ensembl$type=="transcript",]))
table(hg38_gtf_ensembl$type)

hg38_ensembl_transcript_id_gene_name <- data.frame(transcript_id=hg38_gtf_ensembl[hg38_gtf_ensembl$type=="transcript",]$transcript_id,
                                                   gene_id=hg38_gtf_ensembl[hg38_gtf_ensembl$type=="transcript",]$gene_id,
                                                   gene_name=hg38_gtf_ensembl[hg38_gtf_ensembl$type=="transcript",]$gene_name)
tt_wt(hg38_ensembl_transcript_id_gene_name,"/disk/user_09/reference/annotation/hg38/Homo_sapiens.GRCh38.108.gene_name.txt",col.names = T)
```

## mm39
```{r}
## read gtf
mm39_gtf_gencode <- import("/disk/user_09/reference/annotation/mm39/raw/gencode.vM29.basic.annotation.gtf")
mm39_gtf_ensembl <- import("/disk/user_09/reference/annotation/mm39/Mus_musculus.GRCm39.108.gtf")

head(mm39_gtf_gencode$gene_id)
head(mm39_gtf_ensembl$gene_id)

## overlap between gencode and ensembl
mm39_gtf_gencode_gene_id <- sapply(strsplit(mm39_gtf_gencode$gene_id,split = ".",fixed=T),"[",1)
mm39_gtf_gencode_gene_id <- mm39_gtf_gencode_gene_id[!duplicated(mm39_gtf_gencode_gene_id)]

mm39_gtf_ensembl_gene_id <- mm39_gtf_ensembl$gene_id
mm39_gtf_ensembl_gene_id <- mm39_gtf_ensembl_gene_id[!duplicated(mm39_gtf_ensembl_gene_id)]

venn(list(mm39_gtf_gencode_gene_id,mm39_gtf_ensembl_gene_id))

head(mm39_gtf_gencode$transcript_id)
head(mm39_gtf_ensembl$transcript_id)


mm39_gtf_gencode_transcript_id <- sapply(strsplit(mm39_gtf_gencode$transcript_id,split = ".",fixed=T),"[",1)
mm39_gtf_gencode_transcript_id <- mm39_gtf_gencode_transcript_id[!duplicated(mm39_gtf_gencode_transcript_id)]

mm39_gtf_ensembl_transcript_id <- mm39_gtf_ensembl$transcript_id
mm39_gtf_ensembl_transcript_id <- mm39_gtf_ensembl_transcript_id[!duplicated(mm39_gtf_ensembl_transcript_id)]

venn(list(mm39_gtf_gencode_transcript_id,mm39_gtf_ensembl_transcript_id))

## chr modified 
table(mm39_gtf_gencode[mm39_gtf_gencode$type=="gene",]@seqnames)
table(mm39_gtf_ensembl[mm39_gtf_ensembl$type=="gene",]@seqnames)
chr_name <- c(seq(1,19),"X","Y","MT")
mm39_gtf_ensembl_chr <- mm39_gtf_ensembl[mm39_gtf_ensembl@seqnames %in% chr_name,]
table(mm39_gtf_ensembl_chr[mm39_gtf_ensembl_chr$type=="gene",]@seqnames)
tail(mm39_gtf_ensembl_chr[mm39_gtf_ensembl_chr$type=="gene",]@seqnames)

mm39_gtf_ensembl_chr@seqnames <- Rle(factor(paste0("chr",mm39_gtf_ensembl_chr@seqnames),levels = paste0("chr",c(seq(1,19),"X","Y","MT"))))
tail(mm39_gtf_ensembl_chr[mm39_gtf_ensembl_chr$type=="gene",]@seqnames)
mm39_gtf_ensembl_chr@seqnames <- Rle(factor(ifelse(mm39_gtf_ensembl_chr@seqnames=="chrMT","chrM",mm39_gtf_ensembl_chr@seqnames),levels = paste0("chr",c(seq(1,19),"X","Y","M"))))
tail(mm39_gtf_ensembl_chr[mm39_gtf_ensembl_chr$type=="gene",]@seqnames)

export(mm39_gtf_ensembl_chr,"~/reference/annotation/mm39/Mus_musculus.GRCm39.108.chr.gtf","gtf")
```

```{bash}
cd /disk/user_09/reference/annotation/mm39
sed 's/three_prime_utr/3UTR/g' Mus_musculus.GRCm39.108.chr.gtf | sed 's/five_prime_utr/5UTR/g' > Mus_musculus.GRCm39.108.chr_UTR.gtf 
```

```{r basic pc_lnc}

mm39_gtf_ensembl_chr <- import("~/reference/annotation/mm39/Mus_musculus.GRCm39.108.chr_UTR.gtf")

table(mm39_gtf_ensembl_chr$type)

mm39_gtf_ensembl_basic <- mm39_gtf_ensembl_chr[mm39_gtf_ensembl_chr$gene_id %in% mm39_gtf_gencode_gene_id,]
mm39_gtf_ensembl_basic <- mm39_gtf_ensembl_basic[mm39_gtf_ensembl_basic$transcript_id %in% mm39_gtf_gencode_transcript_id,]

export(mm39_gtf_ensembl_basic,"~/reference/annotation/mm39/Mus_musculus.GRCm39.108.basic.gtf","gtf")

mm39_gtf_ensembl_basic_pc_lnc <- mm39_gtf_ensembl_basic[mm39_gtf_ensembl_basic$gene_biotype %in% c("lncRNA","protein_coding"),]
export(mm39_gtf_ensembl_basic_pc_lnc,"~/reference/annotation/mm39/Mus_musculus.GRCm39.108.basic.pc_lnc.gtf","gtf")

mm39_gtf_ensembl_basic_pc <- mm39_gtf_ensembl_basic[mm39_gtf_ensembl_basic$gene_biotype %in% c("protein_coding"),]
export(mm39_gtf_ensembl_basic_pc,"~/reference/annotation/mm39/Mus_musculus.GRCm39.108.basic.pc.gtf","gtf")

```

### gencode

```{r gtf_basic_pc_lncRNA}
gtf_all <- import("~/reference/annotation/mm39/raw/gencode.vM29.annotation.gtf")
gtf_basic <- import("~/reference/annotation/mm39/raw/gencode.vM29.basic.annotation.gtf")

gtf_all[gtf_all$type=="gene",]
gtf_basic[gtf_basic$type=="gene",]

table(gtf_all[gtf_all$type=="gene",]$gene_type)
gtf_basic[gtf_basic$type=="gene",]

gtf_all[gtf_all$type=="transcript",]
gtf_basic[gtf_basic$type=="transcript",]

table(gtf_all[gtf_all$type=="transcript",]$transcript_type)
table(gtf_basic[gtf_basic$type=="transcript",]$transcript_type)

gtf_basic_pc_lnc <- gtf_basic[gtf_basic$gene_type %in% c("lncRNA","protein_coding"),]

table(gtf_basic_pc_lnc[gtf_basic_pc_lnc$type=="transcript",]$transcript_type)

export(gtf_basic_pc_lnc,"~/reference/annotation/mm39/raw/gencode.vM29.basic.annotation.pc_lnc.gtf","gtf")

```

