---
title: "r01_DEG_mRNA_p0p5p10rp2"
author: "Tang Li"
date: '2023-2-11'
output: html_document
---

```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE)
library(DESeq2)
library(GEOquery)
library(GenomicFeatures)
library(BiocParallel)
library(GenomicAlignments)
library(rtracklayer)
library(NMF)
library(parallel)
library(Rsubread)
library(RUVSeq)
library(stringr)
library(data.table)
library(ggplot2)
library(ggprism)
library(RColorBrewer)
library(ttFunctions)
library(ggthemes)
library(ggsci)
library(ggpubr)
library(ComplexHeatmap)
library(sva)
library(venn)
options(stringsAsFactors = FALSE)
library(ggrepel)
library(clusterProfiler)
library(org.Mm.eg.db)
library(biomaRt)
library(enrichplot)
library(patchwork)

# ~/Data/01_TC1/01_mRNA_expression/01_p0p5p10rp2/01_p0p5p10rp2.Rproj

require("knitr")
opts_knit$set(root.dir = file.path("~/Data/01_TC1/01_mRNA_expression/01_p0p5p10rp2/"))

```

```{r}
#generage p6_siNudt21 bam list
dir1='/disk/user_08/Data/TC1-planB/35_Nudt21_EB/03_hisat2_mapping/03_bam_merge/'
sample_name1 <- paste(rep(c("p6"),each=6),rep(c("siNC","siNudt21_1","siNudt21_3"),each=2),
                      rep(c("rep1","rep2"),times=3),
                      sep = "_")

#generate siVirma siM3 M3IN bam list
dir2 <- "/disk/user_09/Data/01_TC1/12_TC1_P16_si/03_hisat2_mapping/03_bam_merge/"
sample_name2 <- paste(rep(c("Ctrl","M3IN","siNC","siM3","siVirma"),each=2),
                      rep(c("rep1","rep2"),times=5),
                      sep = "_")

#generate Nudt21 bam list
dir3 <- "/disk/user_09/Data/01_TC1/24_mito_ribo_Nudt21/03_hisat2_mapping/03_bam_merge/"
sample_name3 <- c("P8_Ctrl_Ctrl","P8_M3IN_Ctrl",
                  "P8_M3IN_At-c1","P8_M3IN_At-c2",
                  "P8_M3IN_CCCP-c1","P8_M3IN_CCCP-c2",
                  "P8_M3IN_CX-c1","P8_M3IN_CX-c2",
                  "P8_M3IN_Rbin-c1","P8_M3IN_Rbin-c2",
                  "P14_siNC-1","P14_siNudt21","P14_siM3",
                  "P14_siNC-2","P14_siVirma",
                  "P0_siNudt21_Ctrl","P0_siNudt21_dox",
                  "P2_siNudt21_Ctrl","P2_siNudt21_dox",
                  "P4_siNudt21_Ctrl","P4_siNudt21_dox",
                  "P5_siNudt21_Ctrl","P5_siNudt21_dox")

#generate mRNA bam list
dir4 <- "/disk/user_08/Data/TC1-planB/11_mRNA/05_bam_hisat2/01_bam_sorted/11_bam_merge/"
sample_name4 <- paste(rep(c("p0","p5","p10","rp2"),each=2),
                      rep(c("input"),each=8),
                      rep(c("rep1","rep2"),times=4),
                      sep = "_")

fls <- c(file.path(dir1,paste0(sample_name1,'.bam')),
         file.path(dir2,paste0(sample_name2,'.bam')),
         file.path(dir3,paste0(sample_name3,'.bam')),
         file.path(dir4,paste0(sample_name4,'.bam')))
file.exists(fls)
```

## featurecount

```{bash}
#### 由于mRNA p0p5p10的bam文件有问题(R语言featureCounts时闪退)，所以在bash上跑featureCounts
cd /disk/user_09/Data/01_TC1/01_mRNA_expression/01_p0p5p10rp2
conda activate subread
featureCounts -a /disk/user_09/reference/annotation/mm39/raw/gencode.vM29.basic.annotation.pc.gtf \
  -o express_matrix_featureCounts_gencode.vM29.basic.annotation.pc_exon_minOverlap30.txt \
  -F GTF -t exon -g 'gene_id' --extraAttributes 'gene_name','gene_type' \
  --minOverlap 30 -s 2 -p -T 50 \
  /disk/user_08/Data/TC1-planB/35_Nudt21_EB/03_hisat2_mapping/03_bam_merge/p6_siNC_rep1.bam \
  /disk/user_08/Data/TC1-planB/35_Nudt21_EB/03_hisat2_mapping/03_bam_merge/p6_siNC_rep2.bam \
  /disk/user_08/Data/TC1-planB/35_Nudt21_EB/03_hisat2_mapping/03_bam_merge/p6_siNudt21_1_rep1.bam \
  /disk/user_08/Data/TC1-planB/35_Nudt21_EB/03_hisat2_mapping/03_bam_merge/p6_siNudt21_1_rep2.bam \
  /disk/user_08/Data/TC1-planB/35_Nudt21_EB/03_hisat2_mapping/03_bam_merge/p6_siNudt21_3_rep1.bam \
  /disk/user_08/Data/TC1-planB/35_Nudt21_EB/03_hisat2_mapping/03_bam_merge/p6_siNudt21_3_rep2.bam \
  /disk/user_09/Data/01_TC1/12_TC1_P16_si/03_hisat2_mapping/03_bam_merge/Ctrl_rep1.bam \
  /disk/user_09/Data/01_TC1/12_TC1_P16_si/03_hisat2_mapping/03_bam_merge/Ctrl_rep2.bam \
  /disk/user_09/Data/01_TC1/12_TC1_P16_si/03_hisat2_mapping/03_bam_merge/M3IN_rep1.bam \
  /disk/user_09/Data/01_TC1/12_TC1_P16_si/03_hisat2_mapping/03_bam_merge/M3IN_rep2.bam \
  /disk/user_09/Data/01_TC1/12_TC1_P16_si/03_hisat2_mapping/03_bam_merge/siNC_rep1.bam \
  /disk/user_09/Data/01_TC1/12_TC1_P16_si/03_hisat2_mapping/03_bam_merge/siNC_rep2.bam \
  /disk/user_09/Data/01_TC1/12_TC1_P16_si/03_hisat2_mapping/03_bam_merge/siM3_rep1.bam \
  /disk/user_09/Data/01_TC1/12_TC1_P16_si/03_hisat2_mapping/03_bam_merge/siM3_rep2.bam \
  /disk/user_09/Data/01_TC1/12_TC1_P16_si/03_hisat2_mapping/03_bam_merge/siVirma_rep1.bam \
  /disk/user_09/Data/01_TC1/12_TC1_P16_si/03_hisat2_mapping/03_bam_merge/siVirma_rep2.bam \
  /disk/user_09/Data/01_TC1/24_mito_ribo_Nudt21/03_hisat2_mapping/03_bam_merge/P8_Ctrl_Ctrl.bam \
 /disk/user_09/Data/01_TC1/24_mito_ribo_Nudt21/03_hisat2_mapping/03_bam_merge/P8_M3IN_Ctrl.bam \
  /disk/user_09/Data/01_TC1/24_mito_ribo_Nudt21/03_hisat2_mapping/03_bam_merge/P8_M3IN_At-c1.bam \
  /disk/user_09/Data/01_TC1/24_mito_ribo_Nudt21/03_hisat2_mapping/03_bam_merge/P8_M3IN_At-c2.bam \
  /disk/user_09/Data/01_TC1/24_mito_ribo_Nudt21/03_hisat2_mapping/03_bam_merge/P8_M3IN_CCCP-c1.bam \
  /disk/user_09/Data/01_TC1/24_mito_ribo_Nudt21/03_hisat2_mapping/03_bam_merge/P8_M3IN_CCCP-c2.bam \
  /disk/user_09/Data/01_TC1/24_mito_ribo_Nudt21/03_hisat2_mapping/03_bam_merge/P8_M3IN_CX-c1.bam \
  /disk/user_09/Data/01_TC1/24_mito_ribo_Nudt21/03_hisat2_mapping/03_bam_merge/P8_M3IN_CX-c2.bam \
  /disk/user_09/Data/01_TC1/24_mito_ribo_Nudt21/03_hisat2_mapping/03_bam_merge/P8_M3IN_Rbin-c1.bam \
  /disk/user_09/Data/01_TC1/24_mito_ribo_Nudt21/03_hisat2_mapping/03_bam_merge/P8_M3IN_Rbin-c2.bam \
  /disk/user_09/Data/01_TC1/24_mito_ribo_Nudt21/03_hisat2_mapping/03_bam_merge/P14_siNC-1.bam \
  /disk/user_09/Data/01_TC1/24_mito_ribo_Nudt21/03_hisat2_mapping/03_bam_merge/P14_siNudt21.bam \
  /disk/user_09/Data/01_TC1/24_mito_ribo_Nudt21/03_hisat2_mapping/03_bam_merge/P14_siM3.bam \
  /disk/user_09/Data/01_TC1/24_mito_ribo_Nudt21/03_hisat2_mapping/03_bam_merge/P14_siNC-2.bam \
  /disk/user_09/Data/01_TC1/24_mito_ribo_Nudt21/03_hisat2_mapping/03_bam_merge/P14_siVirma.bam \
  /disk/user_09/Data/01_TC1/24_mito_ribo_Nudt21/03_hisat2_mapping/03_bam_merge/P0_siNudt21_Ctrl.bam \
  /disk/user_09/Data/01_TC1/24_mito_ribo_Nudt21/03_hisat2_mapping/03_bam_merge/P0_siNudt21_dox.bam \
  /disk/user_09/Data/01_TC1/24_mito_ribo_Nudt21/03_hisat2_mapping/03_bam_merge/P2_siNudt21_Ctrl.bam \
  /disk/user_09/Data/01_TC1/24_mito_ribo_Nudt21/03_hisat2_mapping/03_bam_merge/P2_siNudt21_dox.bam \
  /disk/user_09/Data/01_TC1/24_mito_ribo_Nudt21/03_hisat2_mapping/03_bam_merge/P4_siNudt21_Ctrl.bam \
  /disk/user_09/Data/01_TC1/24_mito_ribo_Nudt21/03_hisat2_mapping/03_bam_merge/P4_siNudt21_dox.bam \
  /disk/user_09/Data/01_TC1/24_mito_ribo_Nudt21/03_hisat2_mapping/03_bam_merge/P5_siNudt21_Ctrl.bam \
  /disk/user_09/Data/01_TC1/24_mito_ribo_Nudt21/03_hisat2_mapping/03_bam_merge/P5_siNudt21_dox.bam \
  /disk/user_08/Data/TC1-planB/11_mRNA/05_bam_hisat2/01_bam_sorted/11_bam_merge/p0_input_rep1.bam \
  /disk/user_08/Data/TC1-planB/11_mRNA/05_bam_hisat2/01_bam_sorted/11_bam_merge/p0_input_rep2.bam \
  /disk/user_08/Data/TC1-planB/11_mRNA/05_bam_hisat2/01_bam_sorted/11_bam_merge/p5_input_rep1.bam \
  /disk/user_08/Data/TC1-planB/11_mRNA/05_bam_hisat2/01_bam_sorted/11_bam_merge/p5_input_rep2.bam \
  /disk/user_08/Data/TC1-planB/11_mRNA/05_bam_hisat2/01_bam_sorted/11_bam_merge/p10_input_rep1.bam \
  /disk/user_08/Data/TC1-planB/11_mRNA/05_bam_hisat2/01_bam_sorted/11_bam_merge/p10_input_rep2.bam \
  /disk/user_08/Data/TC1-planB/11_mRNA/05_bam_hisat2/01_bam_sorted/11_bam_merge/rp2_input_rep1.bam \
  /disk/user_08/Data/TC1-planB/11_mRNA/05_bam_hisat2/01_bam_sorted/11_bam_merge/rp2_input_rep2.bam \
1> express_matrix_featureCounts_gencode.vM29.basic.annotation.pc_exon_minOverlap30.log 2>&1
```

```{r featurecount_result}
fc_result <- as.data.frame(fread("01_featureCounts/express_matrix_featureCounts_gencode.vM29.basic.annotation.pc_exon_minOverlap30.txt"))

fc_result$Geneid <- str_sub(fc_result$Geneid,1,18)

rownames(fc_result) <- fc_result$Geneid

fc_result_count <- fc_result[,9:55]
colnames(fc_result_count) <- c(sample_name1,sample_name2,sample_name3,sample_name4)

fc_result_annotation <- fc_result[,6:8]

tt_wt(fc_result_count,file.path("01_featureCounts/express_matrix_count.txt"),col.names = T,row.names = T) 
tt_wt(fc_result_annotation,file.path("01_featureCounts/express_matrix_annotation.txt"),col.names = T,row.names = T) 
```

```{r filt genes with duplicated gene name}

table(duplicated(fc_result_annotation$gene_name))

fc_result_count <- fc_result_count[!duplicated(fc_result_annotation$gene_name),]
fc_result_annotation <- fc_result_annotation[rownames(fc_result_count),]

rownames(fc_result_count) <- fc_result_annotation$gene_name
rownames(fc_result_annotation) <- fc_result_annotation$gene_name

tt_wt(fc_result_count,file.path("01_featureCounts/express_matrix_count_rowname_modified.txt"),col.names = T,row.names = T) 
tt_wt(fc_result_annotation,file.path("01_featureCounts/express_matrix_annotation_rowname_modified.txt"),col.names = T,row.names = T) 
```

```{R marker gene table}
marker_all <- read.table("~/Data/01_TC1/00_gene_set/all_20230213_p0p10rp2//all")$V1
marker_toti <- read.table("~/Data/01_TC1/00_gene_set/all_20230213_p0p10rp2//toti")$V1
marker_pluri <- read.table("~/Data/01_TC1/00_gene_set/all_20230213_p0p10rp2//pluri")$V1
marker_TE_TSC <- read.table("~/Data/01_TC1/00_gene_set/all_20230213_p0p10rp2//TE_TSC")$V1
marker_PrE_XEN <- read.table("~/Data/01_TC1/00_gene_set/all_20230213_p0p10rp2//PrE_XEN")$V1
marker_APA <- read.table("~/Data/01_TC1/00_gene_set/all_20230213_p0p10rp2//APA")$V1
marker_APAperturb <- read.table("~/Data/01_TC1/00_gene_set/all/APA_perturb")$V1

table(marker_all %in% rownames(express_matrix))
marker_all[!marker_all %in% rownames(express_matrix)]
marker_all <- marker_all[marker_all %in% rownames(express_matrix)]
```

### DEG for p0p5p10rp2

```{r betweenLaneNormalization & expression filter10}

sample_name <- paste(rep(c("p0","p5","p10","rp2"),each=2),rep(c("rep1","rep2"),times=4),sep = "_")
express_matrix <- fc_result_count[,40:47]
colnames(express_matrix) <- sample_name
meta_data <- data.frame(rep = rep(c("rep1","rep2"),times=4),
                        stage = rep(c("p0","p5","p10","rp2"),each=2),
                        row.names = sample_name,
                        sample_name = sample_name,
                        stringsAsFactors = TRUE)
meta_data

x <- meta_data$stage
set <- newSeqExpressionSet(as.matrix(express_matrix),
                           phenoData = data.frame(x, row.names=colnames(express_matrix)))
set

colors <- brewer.pal(18, "Paired")
plotRLE(set, outline=FALSE, ylim=c(-2, 2), col=colors[x])
plotPCA(set, col=colors[x], cex=1.2)

set <- betweenLaneNormalization(set, which="upper")
plotRLE(set, outline=FALSE, ylim=c(-2, 2), col=colors[x])
plotPCA(set, col=colors[x], cex=1.2)

express_matrix <- set@assayData$normalizedCounts

dir.create("02_express_matrix")
tt_wt(express_matrix,"02_express_matrix/express_matrix.txt",row.names=T,col.names = T)
# 对表达量进行筛选
filter <- rowSums(express_matrix>10)>=2
table(filter)
express_matrix_filter10 <- express_matrix[filter,]

colnames(express_matrix_filter10)
tt_wt(express_matrix_filter10,"02_express_matrix/express_matrix_filter10.txt",row.names=T,col.names = T)
```

```{r get tpm_matrix}
#convert to TPM

tpm_matrix <- do.call(cbind,base::lapply(colnames(express_matrix),FUN = function(x){
  counts <- express_matrix[,x]
  effLen <- fc_result_annotation[rownames(express_matrix),"Length"]
  rate <- log(counts) - log(effLen)
  denom <- log(sum(exp(rate)))
  return(exp(rate-denom+log(1e6)))
}))

rownames(tpm_matrix) <- rownames(express_matrix)
colnames(tpm_matrix) <- colnames(express_matrix)
table(colSums(tpm_matrix))

write.table(tpm_matrix,file="02_express_matrix/tpm_matrix.txt",quote = F,sep = "\t")
```

```{r marker barplot}
tt_barplot <- function(gene_name){
  temp <- data.frame(group=factor(c("p0","p0","p10","p10","rp2","rp2")),
                   tpm=tpm_matrix[gene_name,c(1,2,5:8)])
  p1 <- ggplot(temp, aes(x = group, y = tpm, fill=group)) +
  geom_bar(stat = "summary",fun=mean,position="dodge",width =0.8,show.legend = F,size=2)+ #绘制柱状图
  geom_point(show.legend = F,position = position_jitter(width = 0.1))+
  stat_summary(geom = "errorbar",fun.data = 'mean_sdl', width = 0.3,show.legend = F)+#误差棒
  labs(x=NULL,y=paste0(gene_name," tpm"))+#标题
  theme_prism(palette = "floral",
              base_fontface = "plain", # 字体样式，可选 bold, plain, italic
              base_family = "sans", # 字体格式，可选 serif, sans, mono, Arial等
              base_size = 16,  # 图形的字体大小
              base_line_size = 0.8, # 坐标轴的粗细
              axis_text_angle = 0)+ # 可选值有 0，45，90，270
  theme(plot.title = element_text(size=16,hjust = -0.5))+
  scale_fill_prism(palette = "floral")+#使用ggprism包修改颜色
   scale_y_continuous(expand = c(0, 0))
  print(p1)
}

dir.create("08_marker")
pdf("08_marker/01_APA_barplot.pdf",width = 3.5,height = 4)
for (i in marker_APA[marker_APA %in% rownames(tpm_matrix)]){
  tt_barplot(i)
}
tt_barplot("Gapdh")
tt_barplot("Actb")
dev.off()

pdf("08_marker/02_toti_barplot.pdf",width = 3.5,height = 4)
for (i in marker_toti[marker_toti %in% rownames(tpm_matrix)]){
  tt_barplot(i)
}
dev.off()

pdf("08_marker/03_pluri_barplot.pdf",width = 3.5,height = 4)
for (i in marker_pluri[marker_pluri %in% rownames(tpm_matrix)]){
  tt_barplot(i)
}
dev.off()

pdf("08_marker/04_TE_TSC_barplot.pdf",width = 3.5,height = 4)
for (i in marker_TE_TSC[marker_TE_TSC %in% rownames(tpm_matrix)]){
  tt_barplot(i)
}
dev.off()

pdf("08_marker/05_PrE_XEN_barplot.pdf",width = 3.5,height = 4)
for (i in marker_PrE_XEN[marker_PrE_XEN %in% rownames(tpm_matrix)]){
  tt_barplot(i)
}
dev.off()

pdf("08_marker/06_APAperturb_barplot.pdf",width = 3.5,height = 4)
for (i in marker_APAperturb[marker_APAperturb %in% rownames(tpm_matrix)]){
  tt_barplot(i)
}
tt_barplot("Gapdh")
tt_barplot("Actb")
dev.off()

```

```{r marker heatmap}
group <- c("toti","pluri","PrE_XEN","TE_TSC","APAperturb")

for (i in group){
  temp <- t(scale(t(tpm_matrix[rownames(tpm_matrix) %in% get(paste0("marker_",i)),c(1,2,5:6)])))
  temp <- na.omit(temp)
  heatmap <- Heatmap(matrix = temp,name="z-score",
                                show_column_names = TRUE,show_row_names=T,
                                cluster_rows = TRUE,cluster_columns = F,
                                width = unit(4,'inches'),
                                col=circlize::colorRamp2(c(-1.18,0,1.18), 
                                                         colorRampPalette(c("#4875aa", "white", "#d7604c"))(3)))


pdf(paste0("08_marker/08_heatmap_cluster_col_",i,".pdf"))
heatmap <- draw(heatmap)
dev.off()

}


temp <- t(scale(t(tpm_matrix[rownames(tpm_matrix) %in% c(marker_toti,marker_pluri),c(1,2,5:6)])))
temp <- na.omit(temp)
heatmap <- Heatmap(matrix = temp,name="z-score",
                                show_column_names = TRUE,show_row_names=T,
                                cluster_rows = TRUE,cluster_columns = F,
                                width = unit(4,'inches'),
                                col=circlize::colorRamp2(c(-1.18,0,1.18), 
                                                         colorRampPalette(c("#4875aa", "white", "#d7604c"))(3)))


pdf("08_marker/08_heatmap_cluster_col_toti_pluri.pdf",height = 13)
heatmap <- draw(heatmap)
dev.off()

```

## DESeq2
### PCA
```{r create dir}
dir.create("03_PCA")
```

```{r PCA - all gene - p0p5p10rp2}

meta_data 
colnames(express_matrix)

dds <- DESeqDataSetFromMatrix(countData = express_matrix,colData = meta_data,design = ~ stage)

#pca plot

rlog <- rlog(dds,blind = FALSE)

pcaData <- plotPCA(rlog,intgroup = c("stage"),returnData = TRUE,ntop=500)
pcaData
percentVar <- round(100 * attr(pcaData,"percentVar"))
percentVar

p1 <- ggplot(pcaData,aes(x = PC1, y = PC2, color=stage)) +
  geom_point(size=3) +
  #geom_label(aes(label=stage,vjust=1))+
  xlab(paste0("PC1:",percentVar[1],"% variance")) +
  ylab(paste0("PC2:",percentVar[2],"% variance")) +
  #xlim(-15,15)+
  #ylim(-15,10)+
  coord_fixed(ratio=percentVar[2]/percentVar[1]) +
  ggtitle("PCA plot mRNA rlog")+
  theme_bw()
  #scale_color_prism(palette = "floral")
  #scale_color_manual(values = c("M3IN_0h" = "#facf87", "M3IN_4h" = "#d1d1ed", "M3IN_12h" = "#b9dbf4"))


pdf("03_PCA/01_p0p5p10rp2_allgene_rlog.pdf",width = 8,height = 7)
print(p1)
dev.off()

vst <- vst(dds,blind = FALSE)

pcaData <- plotPCA(vst,intgroup = c("stage"),returnData = TRUE,ntop=500)
pcaData
percentVar <- round(100 * attr(pcaData,"percentVar"))
percentVar

p1 <- ggplot(pcaData,aes(x = PC1, y = PC2, color=stage)) +
  geom_point(size=3) +
  #geom_label(aes(label=stage,vjust=1))+
  xlab(paste0("PC1:",percentVar[1],"% variance")) +
  ylab(paste0("PC2:",percentVar[2],"% variance")) +
  #xlim(-15,15)+
  #ylim(-15,10)+
  coord_fixed(ratio=percentVar[2]/percentVar[1]) +
  ggtitle("PCA plot mRNA vst")+
  theme_bw()
  #scale_color_prism(palette = "floral")
  #scale_color_manual(values = c("M3IN_0h" = "#facf87", "M3IN_4h" = "#d1d1ed", "M3IN_12h" = "#b9dbf4"))


pdf("03_PCA/01_p0p5p10rp2_allgene_vst.pdf",width = 8,height = 7)
print(p1)
dev.off()
```

```{r PCA - marker gene - p0p5p10rp2}

meta_data 
colnames(express_matrix)

dds <- DESeqDataSetFromMatrix(countData = express_matrix[marker_all,],colData = meta_data,design = ~ stage)

#pca plot

rlog <- rlog(dds,blind = FALSE)

pcaData <- plotPCA(rlog,intgroup = c("stage"),returnData = TRUE,ntop=500)
pcaData
percentVar <- round(100 * attr(pcaData,"percentVar"))
percentVar

p1 <- ggplot(pcaData,aes(x = PC1, y = PC2, color=stage)) +
  geom_point(size=3) +
  #geom_label(aes(label=stage,vjust=1))+
  xlab(paste0("PC1:",percentVar[1],"% variance")) +
  ylab(paste0("PC2:",percentVar[2],"% variance")) +
  #xlim(-15,15)+
  #ylim(-15,10)+
  coord_fixed(ratio=percentVar[2]/percentVar[1]) +
  ggtitle("PCA plot mRNA rlog")+
  theme_bw()
  #scale_color_prism(palette = "floral")
  #scale_color_manual(values = c("M3IN_0h" = "#facf87", "M3IN_4h" = "#d1d1ed", "M3IN_12h" = "#b9dbf4"))


pdf("03_PCA/02_p0p5p10rp2_allmarker_rlog.pdf",width = 8,height = 7)
print(p1)
dev.off()
```

### for DEG analysis
```{r create dds object}

dds <- DESeqDataSetFromMatrix(countData = express_matrix_filter10,colData = meta_data,design = ~ stage)

dir.create("04_DEG_result")
dir.create("05_DEG_heatmap")
dir.create("06_DEG_volcano")
```

```{r DEG analysis}

#DEG
dds <- DESeq(dds)
resultsNames(dds)

res_p0_p10 <- tt_DEG_result(dds,"p0","p10","04_DEG_result")
res_p10_rp2 <- tt_DEG_result(dds,"p10","rp2","04_DEG_result")
res_p0_rp2 <- tt_DEG_result(dds,"p0","rp2","04_DEG_result")


```

## filter1
```{r DEG result filter1}
#filter1

res_p0_p10_filter_1 <- tt_DEG_filter(res_p0_p10,"p0","p10","04_DEG_result",lfcThreshold = 1)
res_p10_rp2_filter_1 <- tt_DEG_filter(res_p10_rp2,"p10","rp2","04_DEG_result",lfcThreshold = 1)
res_p0_rp2_filter_1 <- tt_DEG_filter(res_p0_rp2,"p0","rp2","04_DEG_result",lfcThreshold = 1)

```

```{r DEG heatmap filter1}

tt_DEG_heatmap(res_p0_p10_filter_1,"p0","p10",tpm_matrix,meta_data,save_dir = "05_DEG_heatmap/",lfcThreshold = 1)
tt_DEG_heatmap(res_p10_rp2_filter_1,"p10","rp2",tpm_matrix,meta_data,save_dir = "05_DEG_heatmap/",lfcThreshold = 1)
tt_DEG_heatmap(res_p0_rp2_filter_1,"p0","rp2",tpm_matrix,meta_data,save_dir = "05_DEG_heatmap/",lfcThreshold = 1)

### cluster col
tt_DEG_heatmap(res_p0_p10_filter_1,"p0","p10",tpm_matrix,meta_data,save_dir = "05_DEG_heatmap/",lfcThreshold = 1,cluster_columns = T,name_suffix = "_cluster_col")
tt_DEG_heatmap(res_p10_rp2_filter_1,"p10","rp2",tpm_matrix,meta_data,save_dir = "05_DEG_heatmap/",lfcThreshold = 1,cluster_columns = T,name_suffix = "_cluster_col")
tt_DEG_heatmap(res_p0_rp2_filter_1,"p0","rp2",tpm_matrix,meta_data,save_dir = "05_DEG_heatmap/",lfcThreshold = 1,cluster_columns = T,name_suffix = "_cluster_col")

```

```{r DEG volcano filter1}
tt_DEG_volcano(res_p0_p10,"p0","p10","06_DEG_volcano/",lfcThreshold =1,pdf_width = 8,pdf_height = 6,padj_limit = 100,name_suffix = "_padj100")
tt_DEG_volcano(res_p10_rp2,"p10","rp2","06_DEG_volcano/",lfcThreshold =1,pdf_width = 8,pdf_height = 6,padj_limit = 100,name_suffix = "_padj100")
tt_DEG_volcano(res_p0_rp2,"p0","rp2","06_DEG_volcano/",lfcThreshold =1,pdf_width = 8,pdf_height = 6,padj_limit = 100,name_suffix = "_padj100")

tt_DEG_volcano_mark(res_p0_p10,"p0","p10","06_DEG_volcano/",lfcThreshold =1,label_genes=marker_APA,pdf_width = 8,pdf_height = 6,padj_limit = 100,name_suffix = "_mark_APA")

tt_DEG_volcano_mark(res_p0_rp2,"p0","rp2","06_DEG_volcano/",lfcThreshold =1,label_genes=marker_APA,pdf_width = 8,pdf_height = 6,padj_limit = 100,name_suffix = "_mark_APA")

tt_DEG_volcano_mark(res_p10_rp2,"p10","rp2","06_DEG_volcano/",lfcThreshold =1,label_genes=marker_APA,pdf_width = 8,pdf_height = 6,padj_limit = 100,name_suffix = "_mark_APA")

tt_DEG_volcano_mark(res_p0_p10,"p0","p10","06_DEG_volcano/",lfcThreshold =1,label_genes=marker_toti,pdf_width = 8,pdf_height = 6,padj_limit = 100,name_suffix = "_mark_toti")
```

```{r DEG GO KEGG}

dir.create("07_DEG_GO")

sn <- c("p0","p10","rp2")
for (i in sn){
  for (j in sn){
    if (exists(paste0("res_",i,"_",j,"_filter_1"))){
      tt_enrichGO(genelist = rownames(get(paste0("res_",i,"_",j,"_filter_1")))[get(paste0("res_",i,"_",j,"_filter_1"))$log2FoldChange>0],groupname = paste0("res_",i,"_",j,"_filter_1_up"),filedir = "07_DEG_GO/")
      tt_enrichGO(genelist = rownames(get(paste0("res_",i,"_",j,"_filter_1")))[get(paste0("res_",i,"_",j,"_filter_1"))$log2FoldChange<0],groupname = paste0("res_",i,"_",j,"_filter_1_down"),filedir = "07_DEG_GO/")
    }
  }
}

```

## filter1 p0p10rp2 DEG heatmap
```{r genes diff in p0p10 & p10rp2}

res_p0_p10_up_genename <- read.table("04_DEG_result/filter_1/res_p0_p10up_genename.txt")$V1
res_p0_p10_down_genename <- read.table("04_DEG_result/filter_1/res_p0_p10down_genename.txt")$V1
res_p10_rp2_up_genename <- read.table("04_DEG_result/filter_1/res_p10_rp2up_genename.txt")$V1
res_p10_rp2_down_genename <- read.table("04_DEG_result/filter_1/res_p10_rp2down_genename.txt")$V1

res_p0p10rp2_UD_genename <- intersect(res_p0_p10_up_genename,res_p10_rp2_down_genename)
res_p0p10rp2_DU_genename <- intersect(res_p0_p10_down_genename,res_p10_rp2_up_genename)
tt_wt(res_p0p10rp2_UD_genename,"04_DEG_result/filter_1/res_p0p10rp2_UD_genename.txt")
tt_wt(res_p0p10rp2_DU_genename,"04_DEG_result/filter_1/res_p0p10rp2_DU_genename.txt")

DEG_p0p10rp2 <- unique(c(res_p0p10rp2_UD_genename,res_p0p10rp2_DU_genename))

tpm_matrix_p0p10rp2 <- tpm_matrix[,c(1,2,5:8)]
heatmap <- tpm_matrix_p0p10rp2[DEG_p0p10rp2,] %>% 
  t() %>% scale() %>% t() %>% Heatmap(name="z-score",
                                show_column_names = TRUE,show_row_names=F,
                                show_row_dend = F,
                                cluster_rows =T,cluster_columns = F,
                                width = unit(4,'inches'),
                                border = T,
                                col=circlize::colorRamp2(c(-1.18,0,1.18), 
                                                         colorRampPalette(c("#4875aa", "white", "#d7604c"))(3)))


pdf("04_DEG_result/filter_1/heatmap_p0p10rp2.pdf")
heatmap <- draw(heatmap)
dev.off()

### km改成4
heatmap <- tpm_matrix_p0p10rp2[DEG_p0p10rp2,] %>% 
  t() %>% scale() %>% t() %>% Heatmap(name="z-score",
                                show_column_names = TRUE,show_row_names=F,
                                show_row_dend = F,
                                cluster_rows =T,cluster_columns = F,
                                width = unit(4,'inches'),
                                km=2,
                                border = T,
                                col=circlize::colorRamp2(c(-1.18,0,1.18), 
                                                         colorRampPalette(c("#4875aa", "white", "#d7604c"))(3)))


pdf("04_DEG_result/filter_1/heatmap_p0p10rp2_km2.pdf")
heatmap <- draw(heatmap)
dev.off()

tt_enrichGO(genelist = res_p0p10rp2_UD_genename,filedir = "04_DEG_result/filter_1/",groupname = "p0p10rp2_UD")
tt_enrichGO(genelist = res_p0p10rp2_DU_genename,filedir = "04_DEG_result/filter_1/",groupname = "p0p10rp2_DU")

```

```{r gsea ribo mito}
dir.create("08_marker_gsea")

## ribosome
Ribosome_genes <- read.table("../../00_gene_set/CC_Ribosome")$V1

## ribosome_biogenesis
ribosome_biogenesis_genes <- read.table("~/Data/01_TC1/user_08_TC1/00_gene_set/ribosome_biogenesis")$V1

## mito
mitochondrial_protein_containing_complex_genes <- read.table("~/Data/01_TC1/user_08_TC1/00_gene_set/mitochondrial_protein_containing_complex")$V1
mitochondrion_organization_genes <- read.table("~/Data/01_TC1/user_08_TC1/00_gene_set/mitochondrion_organization")$V1

## extracellular 
extracellular_matrix_organization_genes <- read.table("~/Data/01_TC1/user_08_TC1/00_gene_set/extracellular_matrix_organization")$V1
extracellular_matrix_genes <- read.table("~/Data/01_TC1/user_08_TC1/00_gene_set/extracellular_matrix")$V1

term_ <- c(rep("GO_ribosome_biogenesis",length(ribosome_biogenesis_genes)),
                            rep("GO_Ribosome",length(Ribosome_genes)),
                            rep("GO_mitochondrial_protein_complex",length(mitochondrial_protein_containing_complex_genes)),
                            rep("GO_mitochondrion_organization",length(mitochondrion_organization_genes)),
                            rep("GO_extracellular_matrix_organization",length(extracellular_matrix_organization_genes)),
                            rep("GO_extracellular_matrix",length(extracellular_matrix_genes)))
gene_ <- c(ribosome_biogenesis_genes,Ribosome_genes,
                            mitochondrial_protein_containing_complex_genes,mitochondrion_organization_genes,
                            extracellular_matrix_organization_genes,extracellular_matrix_genes)
msigdb <- data.frame(term=term_,
                     gene=gene_)

res_p0_p10_list <- res_p0_p10$log2FoldChange

names(res_p0_p10_list) <- rownames(res_p0_p10)

res_p0_p10_list <- sort(res_p0_p10_list,decreasing = T)

gsea <- GSEA(res_p0_p10_list,TERM2GENE = msigdb,
             pvalueCutoff = 1,maxGSSize = 2000,minGSSize = 5,
             eps=1e-100,
             nPermSimple = 10000)

# bacth plot
terms <- c("GO_ribosome_biogenesis",
           "GO_Ribosome",
           "GO_mitochondrial_protein_complex",
           "GO_mitochondrion_organization",
           "GO_extracellular_matrix_organization",
           "GO_extracellular_matrix")
# plot
lapply(terms[c(6)], function(x){
  gseaNb(object = gsea,
         geneSetID = x,
         addPval = T,
         subPlot = 2)
}) -> gseaList1

lapply(terms[c(1,3)], function(x){
  gseaNb(object = gsea,
         geneSetID = x,
         addPval = T,
         pvalX = 0.55,pvalY = 0.25,
         subPlot = 2)
}) -> gseaList2

pdf("08_marker_gsea/gsea_p0p10.pdf",width = 4,height = 4)
print(gseaList1)
print(gseaList2)
dev.off()

gseaList <- list(gseaList1[[1]],gseaList2[[1]],gseaList2[[2]])

pdf("08_marker_gsea/gsea_p0p10_combine.pdf",width = 12,height = 4.5)

# combine
cowplot::plot_grid(plotlist = gseaList,ncol = 3,align = 'hv')

dev.off()

```

```{r gsea gseGO}
res_p0_p10_list_ <- as.data.frame(res_p0_p10[,c(2,5)])
res_p0_p10_list_$SYMBOL <- rownames(res_p0_p10_list_)

df <- bitr(rownames(res_p0_p10_list_),
           fromType = "SYMBOL",
           toType = "ENTREZID",
           OrgDb = org.Mm.eg.db)

res_p0_p10_list_ <- merge(res_p0_p10_list_,df,by="SYMBOL")

genelist <- res_p0_p10_list_$log2FoldChange
names(genelist) <- res_p0_p10_list_$ENTREZID
genelist <- sort(genelist,decreasing = T)

GO_kk_entrez <- gseGO(geneList = genelist,
                      ont = "All",
                      OrgDb = org.Mm.eg.db,
                      keyType = "ENTREZID",
                      pvalueCutoff = 0.1)

save(GO_kk_entrez, file = "08_marker_gsea/GSEA_result.RData")

GO_kk <- DOSE::setReadable(GO_kk_entrez,
                           OrgDb = org.Mm.eg.db,
                           keyType = "ENTREZID")

GO_kk_cut <- GO_kk[GO_kk@result$p.adjust<0.01 & abs(GO_kk@result$NES)>1,]

GO_kk_cut_up <- subset(GO_kk_cut,NES>0)
GO_kk_cut_down <- subset(GO_kk_cut,NES<0)

gseaNb(object = GO_kk_entrez,
         geneSetID = "GO:0044391",
         addPval = T,
         subPlot = 2)

pdf("08_marker_gsea/gsea_GO_cut_q0.01.pdf",width = 5,height = 5)

GO_kk_cut_down <- GO_kk_cut_down[order(GO_kk_cut_down$pvalue,decreasing = T),]
GO_kk_cut_down$Description <- factor(GO_kk_cut_down$Description, levels = GO_kk_cut_down$Description)
ggplot(GO_kk_cut_down, aes(x=-log10(pvalue), y=Description, size=abs(NES))) +
  geom_point(aes(color=abs(NES))) +
  #scale_size_continuous(range=c(0.5,2.5)) +
  theme_classic() +
  labs(x="-log10(pValue)", y=NULL)+
  scale_color_gradient(low = "yellow", high = "red")+
  theme(axis.text = element_text(size=10))
dev.off()

pdf("08_marker_gsea/gsea_GO_cut_up_q0.01.pdf",width = 8,height = 7)
GO_kk_cut_up <- GO_kk_cut_up[order(GO_kk_cut_up$pvalue,decreasing = T),]
GO_kk_cut_up$Description <- factor(GO_kk_cut_up$Description, levels = GO_kk_cut_up$Description)
ggplot(GO_kk_cut_up, aes(x=-log10(pvalue), y=Description, size=abs(NES))) +
  geom_point(aes(color=abs(NES))) +
  #scale_size_continuous(range=c(0.5,2.5)) +
  theme_classic() +
  labs(x="-log10_pValue", y=NULL)+
  scale_color_gradient(low = "yellow", high = "red")+
  theme(axis.text = element_text(size=10))
dev.off()
```
