---
title: "r01_m6A_qc"
author: "Tang Li"
date: '2022-11-11'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ChIPpeakAnno)
library(ggplot2)
library(ggpubr)
#devtools::install_github("ricardo-bion/ggradar")
library(ggradar)
require(ggseqlogo)
library(Guitar)
library(GenomicFeatures)
#install.packages('nVennR')
library(nVennR)
library(patchwork)
library(QNB)
library(reshape2)
library(rtracklayer)
library(Rsubread)
library(tidyverse)
#install.packages("Vennerable", repos="http://R-Forge.R-project.org")
library(Vennerable)
library(ttFunctions)


require("knitr")
opts_knit$set(root.dir = "~/Data/01_TC1/21_M3IN_ca_12h/05_m6A_peak/05_m6A_qc/")
```

## replication
```{r peakAnno replication}

common_peak_dir <- "/disk/user_09/Data/01_TC1/21_M3IN_ca_12h/05_m6A_peak/02_bed_modify/"
M3IN_0h_bed1 <- file.path(common_peak_dir,"M3IN_0h_rep1_peaks.bed")
M3IN_0h_bed2 <- file.path(common_peak_dir,"M3IN_0h_rep2_peaks.bed")

M3IN_4h_bed1 <- file.path(common_peak_dir,"M3IN_4h_rep1_peaks.bed")
M3IN_4h_bed2 <- file.path(common_peak_dir,"M3IN_4h_rep2_peaks.bed")

M3IN_12h_bed1 <- file.path(common_peak_dir,"M3IN_12h_rep1_peaks.bed")
M3IN_12h_bed2 <- file.path(common_peak_dir,"M3IN_12h_rep2_peaks.bed")

M3IN_0h_1 <- ChIPpeakAnno::toGRanges(M3IN_0h_bed1, format="BED", header=FALSE)
M3IN_0h_2 <- ChIPpeakAnno::toGRanges(M3IN_0h_bed2, format="BED", header=FALSE)
M3IN_4h_1 <- ChIPpeakAnno::toGRanges(M3IN_4h_bed1, format="BED", header=FALSE)
M3IN_4h_2 <- ChIPpeakAnno::toGRanges(M3IN_4h_bed2, format="BED", header=FALSE)
M3IN_12h_1 <- ChIPpeakAnno::toGRanges(M3IN_12h_bed1, format="BED", header=FALSE)
M3IN_12h_2 <- ChIPpeakAnno::toGRanges(M3IN_12h_bed2, format="BED", header=FALSE)

## must keep the class exactly same as gr1$score, i.e., numeric.
M3IN_0h_1$score <- as.numeric(M3IN_0h_1$score) 
M3IN_0h_2$score <- as.numeric(M3IN_0h_2$score) 
M3IN_4h_1$score <- as.numeric(M3IN_4h_1$score) 
M3IN_4h_2$score <- as.numeric(M3IN_4h_2$score) 
M3IN_12h_1$score <- as.numeric(M3IN_12h_1$score) 
M3IN_12h_2$score <- as.numeric(M3IN_12h_2$score) 

ol_M3IN_0h <- findOverlapsOfPeaks(M3IN_0h_1, M3IN_0h_2)
## add metadata (mean of score) to the overlapping peaks
ol_M3IN_0h <- addMetadata(ol_M3IN_0h, colNames="score", FUN=base::mean) 
ol_M3IN_0h$peaklist[["M3IN_0h_1///M3IN_0h_2"]]

ol_M3IN_4h <- findOverlapsOfPeaks(M3IN_4h_1, M3IN_4h_2)
## add metadata (mean of score) to the overlapping peaks
ol_M3IN_4h <- addMetadata(ol_M3IN_4h, colNames="score", FUN=base::mean) 
ol_M3IN_4h$peaklist[["M3IN_4h_1///M3IN_4h_2"]]

ol_M3IN_12h <- findOverlapsOfPeaks(M3IN_12h_1, M3IN_12h_2)
## add metadata (mean of score) to the overlapping peaks
ol_M3IN_12h <- addMetadata(ol_M3IN_12h, colNames="score", FUN=base::mean) 
ol_M3IN_12h$peaklist[["M3IN_12h_1///M3IN_12h_2"]]

dir.create("01_replication")
pdf("01_replication/01_peak_overlap.pdf")
makeVennDiagram(ol_M3IN_0h, fill=c("#b3dbdc", "#fbb2c9"), # circle fill color
                col=c("#D55E00", "#0072B2"), #circle border color
                cat.col=c("#D55E00", "#0072B2"),# label color, keep same as circle border color
                cex=3,
                cat.cex=2.5,
                margin=0.1) 

makeVennDiagram(ol_M3IN_4h, fill=c("#b3dbdc", "#fbb2c9"), # circle fill color
                col=c("#D55E00", "#0072B2"), #circle border color
                cat.col=c("#D55E00", "#0072B2"),# label color, keep same as circle border color
                cex=3,
                cat.cex=2.5,
                margin=0.1) 

makeVennDiagram(ol_M3IN_12h, fill=c("#b3dbdc", "#fbb2c9"), # circle fill color
                col=c("#D55E00", "#0072B2"), #circle border color
                cat.col=c("#D55E00", "#0072B2"),# label color, keep same as circle border color
                cex=3,
                cat.cex=2.5,
                margin=0.1) 
dev.off()
```

```{r peak overlap between group}
M3IN_0h <- ol_M3IN_0h$peaklist[["M3IN_0h_1///M3IN_0h_2"]]
M3IN_4h <- ol_M3IN_4h$peaklist[["M3IN_4h_1///M3IN_4h_2"]]
M3IN_12h <- ol_M3IN_12h$peaklist[["M3IN_12h_1///M3IN_12h_2"]]

#M3IN_0h_export <- as.data.frame(M3IN_0h)
#write.table(M3IN_0h_export[,c(1,2,3,6,5,5)],"~/LinLong/08_bed_filtered/dedup/M3IN_0h_common_peaks.bed",quote = FALSE,row.names = FALSE,col.names = FALSE,sep="\t")

ol <- findOverlapsOfPeaks(M3IN_0h,M3IN_4h,M3IN_12h)

pdf("01_replication/02_peak_overlap_M3IN_0hM3IN_4hM3IN_12h.pdf")
makeVennDiagram(ol, fill=c("#b3dbdc", "#fbb2c9","#f9be72"), # circle fill color
                col=c("#D55E00", "#0072B2","#c5975a"), #circle border color
                cat.col=c("#D55E00", "#0072B2","#c5975a"),# label color, keep same as circle border color
                cex=3,
                cat.cex=2,
                margin=0.1) 
dev.off()
ol$venn_cnt
venn::venn(x=3,counts=ol$venn_cnt[,4])
venn_data <- Venn(Weight = ol$venn_cnt[,4],SetNames=c("12h","4h","0h"),numberOfSets = 3)

pdf("01_replication/03_peak_overlap_M3IN_0hM3IN_4hM3IN_12h.pdf")
plot(venn_data, show = list(SetLabels = T,Faces = T, DarkMatter = F))
dev.off()

pdf("01_replication/04_peak_overlap_M3IN_0hM3IN_4hM3IN_12h.pdf")
myV <- createVennObj(nSets = 3, sNames = c("0h", "4h", "12h"), sSizes = ol$venn_cnt[,4])
myV <- plotVenn(nVennObj = myV, setColors=c("#0072B2","#D55E00", "#c5975a"),outFile="01_replication/04_peak_overlap_M3IN_0hM3IN_4hM3IN_12h3.svg") 
dev.off()
```

```{r peak anno}

txdb <- makeTxDbFromGFF('/disk/user_09/reference/annotation/mm39/gencode.vM29.annotation.protein_coding.chr.gtf')
annoData <- toGRanges(txdb, format='gene')
annoData[1:2]


overlaps_M3IN_0h <- ol$peaklist[["M3IN_0h"]]
overlaps_M3IN_4h <- ol$peaklist[["M3IN_4h"]]
overlaps_M3IN_12h <- ol$peaklist[["M3IN_12h"]]
overlaps <- ol$peaklist[["M3IN_0h///M3IN_4h///M3IN_12h"]]

# binOverFeature(overlaps, annotationData=annoData,
#                radius=5000, nbins=50, FUN=length, errFun=0,
#                xlab="distance from TSS (bp)", ylab="count", 
#                main="Distribution of aggregated peak numbers around TSS(M3IN_12h)")

## check the genomic element distribution of the duplicates
## the genomic element distribution will indicates the 
## the correlation between duplicates.
dir.create("05_annotation")
pdf("05_annotation/01_peak_ElementDistribution.pdf")
peaks_sep <- GRangesList(M3IN_0h_1=M3IN_0h_1,
                        M3IN_0h_2=M3IN_0h_2,
                        M3IN_4h_1=M3IN_4h_1,
                        M3IN_4h_2=M3IN_4h_2,
                        M3IN_12h_1=M3IN_12h_1,
                        M3IN_12h_2=M3IN_12h_2)
genomicElementDistribution(peaks_sep, 
                           TxDb = txdb,
                           promoterRegion=c(upstream=1000, downstream=100),
                           geneDownstream=c(upstream=0, downstream=4000))

peaks_rep <- GRangesList(M3IN_0h=M3IN_0h,
                         M3IN_4h=M3IN_4h,
                     M3IN_12h=M3IN_12h)
genomicElementDistribution(peaks_rep, 
                           TxDb = txdb,
                           promoterRegion=c(upstream=1000, downstream=100),
                           geneDownstream=c(upstream=0, downstream=4000))
dev.off()

pdf("05_annotation/01_peak_ElementDistribution_circle.pdf")
out_M3IN_0h <- genomicElementDistribution(peaks_rep$M3IN_0h, 
                                         TxDb = txdb,
                                         promoterRegion=c(upstream=1000, downstream=100),
                                         geneDownstream=c(upstream=0, downstream=4000))

out_M3IN_4h <- genomicElementDistribution(peaks_rep$M3IN_4h, 
                                         TxDb = txdb,
                                         promoterRegion=c(upstream=1000, downstream=100),
                                         geneDownstream=c(upstream=0, downstream=4000))

out_M3IN_12h <- genomicElementDistribution(peaks_rep$M3IN_12h, 
                                         TxDb = txdb,
                                         promoterRegion=c(upstream=1000, downstream=100),
                                         geneDownstream=c(upstream=0, downstream=4000))
dev.off()

## check the genomic element distribution for the overlaps
## the genomic element distribution will indicates the 
## the best methods for annotation.
## The percentages in the legend show the percentage of peaks in 
## each category.


# out_M3IN_12h <- genomicElementDistribution(overlaps_M3IN_12h, 
#                                          TxDb = txdb,
#                                          promoterRegion=c(upstream=2000, downstream=500),
#                                          geneDownstream=c(upstream=0, downstream=5000),
#                                          promoterLevel=list(
#                                            # from 5' -> 3', fixed precedence 3' -> 5'
#                                            breaks = c(-2000, -1000, -500, 0, 500),
#                                            labels = c("upstream 1-2Kb", "upstream 0.5-1Kb", 
#                                                       "upstream <500b", "TSS - 500b"),
#                                            colors = c("#FFE5CC", "#FFCA99", 
#                                                       "#FFAD65", "#FF8E32")))
```

```{bash peak_merge_saf}
mkdir 01_peak_merge_saf
for i in {M3IN_0h,M3IN_4h,M3IN_12h}; do
 cat /disk/user_09/Data/01_TC1/21_M3IN_ca_12h/05_m6A_peak/03_bed_merge/03_common_peak_merge/${i}.bed | awk -F "\t" -v OFS="\t" '{print "peak"NR,$1,$2,$3,$4}'> 01_peak_merge_saf/${i}_common_peaks.saf
done
```

```{r reproducibility raw}
bam_dir <- "/disk/user_09/Data/01_TC1/21_M3IN_ca_12h/03_hisat2_mapping/03_bam_merge/"
sample_name <- paste(rep(c("M3IN_0h","M3IN_4h","M3IN_12h"),each=4),
                  rep(c("input","ip"),each=2,times=3),
                  rep(c("rep1","rep2"),times=6),
                 sep="_")

bamfiles <- paste0(bam_dir,sample_name,".bam")

total_Reads <- read.table("/disk/user_09/Data/01_TC1/21_M3IN_ca_12h/03_hisat2_mapping/03_bam_merge/flagstat/flagstat_num_summary.txt")$V1/2

total_Reads <- total_Reads[c(1,2,7,8,3,4,9,10,5,6,11,12)]

## M3IN_0h
M3IN_0h_peak_fc <- featureCounts(files=bamfiles[1:4],
                               annot.ext = '01_replication/01_peak_merge_saf/M3IN_0h_common_peaks.saf',
                               isGTFAnnotationFile = FALSE,
                               useMetaFeatures = FALSE,
                               allowMultiOverlap=FALSE,
                            fracOverlap = 0.5,
                               minMQS = 20, strandSpecific=2,
                               countMultiMappingReads=FALSE,
                               isPairedEnd=TRUE,nthreads=50)

M3IN_0h_counts <- M3IN_0h_peak_fc$counts
colnames(M3IN_0h_counts) <- sample_name[1:4]

colSums(M3IN_0h_counts)
M3IN_0h_cpm <- as.data.frame(t(t(M3IN_0h_counts)/total_Reads[1:4]* 1000000))#参考cpm定义
colSums(M3IN_0h_cpm)

table(rowMeans(M3IN_0h_cpm)>0)
# FALSE  TRUE 
#   181 16870
M3IN_0h_cpm <- M3IN_0h_cpm[rowMeans(M3IN_0h_cpm)>0,]

M3IN_0h_cpm$lfc_rep1 <- log2((M3IN_0h_cpm$`M3IN_0h_ip_rep1`+1)/(M3IN_0h_cpm$`M3IN_0h_input_rep1`+1))
M3IN_0h_cpm$lfc_rep2 <- log2((M3IN_0h_cpm$`M3IN_0h_ip_rep2`+1)/(M3IN_0h_cpm$`M3IN_0h_input_rep2`+1))
summary(M3IN_0h_cpm$lfc_rep1 )
summary(M3IN_0h_cpm$lfc_rep2 )

#### M3IN_4h
M3IN_4h_peak_fc <- featureCounts(files=bamfiles[5:8],
                               annot.ext = '01_replication/01_peak_merge_saf/M3IN_4h_common_peaks.saf',
                               isGTFAnnotationFile = FALSE,
                               useMetaFeatures = FALSE,
                               allowMultiOverlap=FALSE,
                            fracOverlap = 0.5,
                               minMQS = 20, strandSpecific=2,
                               countMultiMappingReads=FALSE,
                               isPairedEnd=TRUE,nthreads=50)

M3IN_4h_counts <- M3IN_4h_peak_fc$counts
colnames(M3IN_4h_counts) <- sample_name[5:8]

colSums(M3IN_4h_counts)
M3IN_4h_cpm <- as.data.frame(t(t(M3IN_4h_counts)/total_Reads[5:8]* 1000000))#参考cpm定义
colSums(M3IN_4h_cpm)

table(rowMeans(M3IN_4h_cpm)>0)
# FALSE  TRUE 
#    53 11802 
M3IN_4h_cpm <- M3IN_4h_cpm[rowMeans(M3IN_4h_cpm)>0,]

M3IN_4h_cpm$lfc_rep1 <- log2((M3IN_4h_cpm$`M3IN_4h_ip_rep1`+1)/(M3IN_4h_cpm$`M3IN_4h_input_rep1`+1))
M3IN_4h_cpm$lfc_rep2 <- log2((M3IN_4h_cpm$`M3IN_4h_ip_rep2`+1)/(M3IN_4h_cpm$`M3IN_4h_input_rep2`+1))
summary(M3IN_4h_cpm$lfc_rep1 )
summary(M3IN_4h_cpm$lfc_rep2 )

#### M3IN_12h
M3IN_12h_peak_fc <- featureCounts(files=bamfiles[9:12],
                               annot.ext = '01_replication/01_peak_merge_saf/M3IN_12h_common_peaks.saf',
                               isGTFAnnotationFile = FALSE,
                               useMetaFeatures = FALSE,
                               allowMultiOverlap=FALSE,
                            fracOverlap = 0.5,
                               minMQS = 20, strandSpecific=2,
                               countMultiMappingReads=FALSE,
                               isPairedEnd=TRUE,nthreads=50)

M3IN_12h_counts <- M3IN_12h_peak_fc$counts
colnames(M3IN_12h_counts) <- sample_name[9:12]

colSums(M3IN_12h_counts)
M3IN_12h_cpm <- as.data.frame(t(t(M3IN_12h_counts)/total_Reads[9:12]* 1000000))#参考cpm定义
colSums(M3IN_12h_cpm)

table(rowMeans(M3IN_12h_cpm)>0)
# FALSE  TRUE 
#    69 11888 
M3IN_12h_cpm <- M3IN_12h_cpm[rowMeans(M3IN_12h_cpm)>0,]

M3IN_12h_cpm$lfc_rep1 <- log2((M3IN_12h_cpm$`M3IN_12h_ip_rep1`+1)/(M3IN_12h_cpm$`M3IN_12h_input_rep1`+1))
M3IN_12h_cpm$lfc_rep2 <- log2((M3IN_12h_cpm$`M3IN_12h_ip_rep2`+1)/(M3IN_12h_cpm$`M3IN_12h_input_rep2`+1))
summary(M3IN_12h_cpm$lfc_rep1 )
summary(M3IN_12h_cpm$lfc_rep2 )


M3IN_0h_rep_point <- ggplot(M3IN_0h_cpm, aes(x = lfc_rep1, y = lfc_rep2)) +
  geom_point(alpha=0.5) + # color="#99CC33"
  geom_smooth(method = "lm",color= "#377EB8")+
  stat_cor(method = "pearson",size=8)+
  labs(x="Replicate 1", y="Replicate 2",title='M3IN_0h')+
  theme_pubr()+
  theme(plot.margin = unit(c(0.5,0.5,0.5,0.5), "cm"))+ #调整与图片边缘的距离
  theme(axis.title.x = element_text(size = 20,face="bold"),axis.title.y = element_text(size = 20,face="bold"),plot.title = element_text(size=25,face="bold", hjust = 0.45) )+
  coord_fixed(ratio=1)

M3IN_4h_rep_point <- ggplot(M3IN_4h_cpm, aes(x = lfc_rep1, y = lfc_rep2)) +
  geom_point(alpha=0.5) + # color="#99CC33"
  geom_smooth(method = "lm",color= "#377EB8")+
  stat_cor(method = "pearson",size=8)+
  labs(x="Replicate 1", y="Replicate 2",title='M3IN_4h')+
  theme_pubr()+
  theme(plot.margin = unit(c(0.5,0.5,0.5,0.5), "cm"))+ #调整与图片边缘的距离
  theme(axis.title.x = element_text(size = 20,face="bold"),axis.title.y = element_text(size = 20,face="bold"),plot.title = element_text(size=25,face="bold", hjust = 0.45) )+
  coord_fixed(ratio=1)

M3IN_12h_rep_point <- ggplot(M3IN_12h_cpm, aes(x = lfc_rep1, y = lfc_rep2)) +
  geom_point(alpha=0.5) + # color="#99CC33"
  geom_smooth(method = "lm",color= "#377EB8")+
  stat_cor(method = "pearson",size=8)+
  labs(x="Replicate 1", y="Replicate 2",title='M3IN_12h')+
  theme_pubr()+
  theme(plot.margin = unit(c(0.5,0.5,0.5,0.5), "cm"))+ #调整与图片边缘的距离
  theme(axis.title.x = element_text(size = 20,face="bold"),axis.title.y = element_text(size = 20,face="bold"),plot.title = element_text(size=25,face="bold", hjust = 0.45) )+
  coord_fixed(ratio=1)
pdf("01_replication/05_raw_reproducibility2.pdf",width = 20)
M3IN_0h_rep_point +M3IN_4h_rep_point+M3IN_12h_rep_point
dev.off()

```

```{r reproducibility spikein}

scale_facter <- data.frame(group=factor(c("M3IN_0h","M3IN_0h","M3IN_4h","M3IN_4h","M3IN_12h","M3IN_12h"),levels=c("M3IN_0h","M3IN_4h","M3IN_12h")),
                      m6A_level=c(1.730130615,1.602430411,0.960775359,1.042563735,1.122506734,1.13664236))

#### M3IN_0h
M3IN_0h_cpm$lfc_rep1_spikein <- log2((M3IN_0h_cpm$`M3IN_0h_ip_rep1`*scale_facter$m6A_level[1]+1)/(M3IN_0h_cpm$`M3IN_0h_input_rep1`+1))
M3IN_0h_cpm$lfc_rep2_spikein <- log2((M3IN_0h_cpm$`M3IN_0h_ip_rep2`*scale_facter$m6A_level[2]+1)/(M3IN_0h_cpm$`M3IN_0h_input_rep2`+1))
summary(M3IN_0h_cpm$lfc_rep1_spikein )
summary(M3IN_0h_cpm$lfc_rep2_spikein )

#### M3IN_4h
M3IN_4h_cpm$lfc_rep1_spikein <- log2((M3IN_4h_cpm$`M3IN_4h_ip_rep1`*scale_facter$m6A_level[3]+1)/(M3IN_4h_cpm$`M3IN_4h_input_rep1`+1))
M3IN_4h_cpm$lfc_rep2_spikein <- log2((M3IN_4h_cpm$`M3IN_4h_ip_rep2`*scale_facter$m6A_level[4]+1)/(M3IN_4h_cpm$`M3IN_4h_input_rep2`+1))
summary(M3IN_4h_cpm$lfc_rep1_spikein )
summary(M3IN_4h_cpm$lfc_rep2_spikein )

#### M3IN_12h
M3IN_12h_cpm$lfc_rep1_spikein <- log2((M3IN_12h_cpm$`M3IN_12h_ip_rep1`*scale_facter$m6A_level[5]+1)/(M3IN_12h_cpm$`M3IN_12h_input_rep1`+1))
M3IN_12h_cpm$lfc_rep2_spikein <- log2((M3IN_12h_cpm$`M3IN_12h_ip_rep2`*scale_facter$m6A_level[6]+1)/(M3IN_12h_cpm$`M3IN_12h_input_rep2`+1))
summary(M3IN_12h_cpm$lfc_rep1_spikein )
summary(M3IN_12h_cpm$lfc_rep2_spikein )

pdf("01_replication/06_spikein_reproducibility.pdf")
ggplot(M3IN_0h_cpm, aes(x = lfc_rep1_spikein, y = lfc_rep2_spikein)) +
  geom_point(alpha=0.5) + # color="#99CC33"
  geom_smooth(method = "lm",color= "#377EB8")+
  stat_cor(method = "pearson",size=8)+
  labs(x="Replicate 1", y="Replicate 2",title='M3IN_0h spikein')+
  theme_pubr()+
  theme(plot.margin = unit(c(0.5,0.5,0.5,0.5), "cm"))+ #调整与图片边缘的距离
  theme(axis.title.x = element_text(size = 20,face="bold"),axis.title.y = element_text(size = 20,face="bold"),plot.title = element_text(size=25,face="bold", hjust = 0.45) ) 

ggplot(M3IN_4h_cpm, aes(x = lfc_rep1_spikein, y = lfc_rep2_spikein)) +
  geom_point(alpha=0.5) + # color="#99CC33"
  geom_smooth(method = "lm",color= "#377EB8")+
  stat_cor(method = "pearson",size=8)+
  labs(x="Replicate 1", y="Replicate 2",title='M3IN_4h spikein')+
  theme_pubr()+
  theme(plot.margin = unit(c(0.5,0.5,0.5,0.5), "cm"))+ #调整与图片边缘的距离
  theme(axis.title.x = element_text(size = 20,face="bold"),axis.title.y = element_text(size = 20,face="bold"),plot.title = element_text(size=25,face="bold", hjust = 0.45) ) 

ggplot(M3IN_12h_cpm, aes(x = lfc_rep1_spikein, y = lfc_rep2_spikein)) +
  geom_point(alpha=0.5) + # color="#99CC33"
  geom_smooth(method = "lm",color= "#377EB8")+
  stat_cor(method = "pearson",size=8)+
  labs(x="Replicate 1", y="Replicate 2",title='M3IN_12h spikein')+
  theme_pubr()+
  theme(plot.margin = unit(c(0.5,0.5,0.5,0.5), "cm"))+ #调整与图片边缘的距离
  theme(axis.title.x = element_text(size = 20,face="bold"),
        axis.title.y = element_text(size = 20,face="bold"),
        plot.title = element_text(size=25,face="bold", hjust = 0.45) ) 

dev.off()

```

```{bash peak_merge_saf}
conda activate bedtools 
cd ~/Data/01_TC1/21_M3IN_ca_12h/05_m6A_peak/04_bed_group_merge
cat ../03_bed_merge/03_common_peak_merge/M3IN_0h.bed ../03_bed_merge/03_common_peak_merge/M3IN_4h.bed \
    | sort -k1,1 -k2,2n \
    | bedtools merge -s -c 6 -o distinct\
    | awk -F "\t" -v OFS="\t" '{{print $1,$2,$3,"peak"NR,".",$4}}' \
    > M3IN_0h_M3IN_4h.bed
    
cat ../03_bed_merge/03_common_peak_merge/M3IN_0h.bed ../03_bed_merge/03_common_peak_merge/M3IN_12h.bed \
    | sort -k1,1 -k2,2n \
    | bedtools merge -s -c 6 -o distinct\
    | awk -F "\t" -v OFS="\t" '{{print $1,$2,$3,"peak"NR,".",$4}}' \
    > M3IN_0h_M3IN_12h.bed

cd ~/Data/01_TC1/21_M3IN_ca_12h/05_m6A_peak/05_m6A_qc/01_replication
mkdir 02_peak_merge_group_saf
for i in {M3IN_0h_M3IN_4h,M3IN_0h_M3IN_12h}; do
 cat ../../04_bed_group_merge/${i}.bed | awk -F "\t" -v OFS="\t" '{print "peak"NR,$1,$2,$3,$4}'> 02_peak_merge_group_saf/${i}_peaks.saf
done
```

```{r peak m6A level change}

scale_facter <- data.frame(group=factor(c("M3IN_0h","M3IN_4h","M3IN_12h"),levels=c("M3IN_0h","M3IN_4h","M3IN_12h")),
                      m6A_level=c(1.66,1,1.13))

## M3IN_0hM3IN_4h
M3IN_0hM3IN_4h_peak_fc <- featureCounts(files=bamfiles[1:8],
                               annot.ext = '/disk/user_09/Data/01_TC1/21_M3IN_ca_12h/05_m6A_peak/05_m6A_qc/01_replication/02_peak_merge_group_saf/M3IN_0h_M3IN_4h_peaks.saf',
                               isGTFAnnotationFile = FALSE,
                               useMetaFeatures = FALSE,
                               allowMultiOverlap=FALSE,
                            fracOverlap = 0.5,
                               minMQS = 20, strandSpecific=2,
                               countMultiMappingReads=FALSE,
                               isPairedEnd=TRUE,nthreads=50)

M3IN_0hM3IN_4h_counts <- M3IN_0hM3IN_4h_peak_fc$counts
colnames(M3IN_0hM3IN_4h_counts) <- sample_name[1:8]

colSums(M3IN_0hM3IN_4h_counts)
M3IN_0hM3IN_4h_cpm <- as.data.frame(t(t(M3IN_0hM3IN_4h_counts)/total_Reads[1:8]* 1000000))#参考cpm定义
colSums(M3IN_0hM3IN_4h_cpm)

table(rowMeans(M3IN_0hM3IN_4h_cpm)>0)
# FALSE  TRUE 
#   183 20204
M3IN_0hM3IN_4h_cpm <- M3IN_0hM3IN_4h_cpm[rowMeans(M3IN_0hM3IN_4h_cpm)>0,]

M3IN_0hM3IN_4h_cpm_mean <- data.frame(M3IN_0h_input=apply(M3IN_0hM3IN_4h_cpm[,1:2],1,mean),
                             M3IN_0h_ip=apply(M3IN_0hM3IN_4h_cpm[,3:4],1,mean),
                             M3IN_4h_input=apply(M3IN_0hM3IN_4h_cpm[,5:6],1,mean),
                             M3IN_4h_ip=apply(M3IN_0hM3IN_4h_cpm[,7:8],1,mean))
### 到底是加1还是0.001还是不加？
M3IN_0hM3IN_4h_cpm_mean$lfc_M3IN_0h_spikein <- log2((M3IN_0hM3IN_4h_cpm_mean$`M3IN_0h_ip`*scale_facter$m6A_level[1]+1)/(M3IN_0hM3IN_4h_cpm_mean$`M3IN_0h_input`+1))
M3IN_0hM3IN_4h_cpm_mean$lfc_M3IN_4h_spikein <- log2((M3IN_0hM3IN_4h_cpm_mean$`M3IN_4h_ip`*scale_facter$m6A_level[2]+1)/(M3IN_0hM3IN_4h_cpm_mean$`M3IN_4h_input`+1))

M3IN_0hM3IN_4h_cpm_mean <- M3IN_0hM3IN_4h_cpm_mean[is.finite(apply(M3IN_0hM3IN_4h_cpm_mean,1,sum)),]
summary(M3IN_0hM3IN_4h_cpm_mean$lfc_M3IN_0h_spikein )
summary(M3IN_0hM3IN_4h_cpm_mean$lfc_M3IN_4h_spikein )

M3IN_0hM3IN_4h_cpm_mean$group <- ifelse(M3IN_0hM3IN_4h_cpm_mean$lfc_M3IN_0h_spikein-M3IN_0hM3IN_4h_cpm_mean$lfc_M3IN_4h_spikein>1,"M3IN_0h_high",ifelse(M3IN_0hM3IN_4h_cpm_mean$lfc_M3IN_4h_spikein-M3IN_0hM3IN_4h_cpm_mean$lfc_M3IN_0h_spikein>1,"M3IN_4h_high","NC"))
M3IN_0hM3IN_4h_cpm_mean_filter <- M3IN_0hM3IN_4h_cpm_mean[(M3IN_0hM3IN_4h_cpm_mean$lfc_M3IN_0h_spikein > 1) | (M3IN_0hM3IN_4h_cpm_mean$lfc_M3IN_4h_spikein > 1),]
table(M3IN_0hM3IN_4h_cpm_mean_filter$group)
# M3IN_0h_high M3IN_4h_high           NC 
#         9706          101         8980


tt_wt(M3IN_0hM3IN_4h_cpm_mean_filter,"02_peak_m6A_level/f01_M3IN_0hM3IN_4h_cpm_mean_filter.txt",row.names = T,col.names = T)

## M3IN_0hM3IN_12h
M3IN_0hM3IN_12h_peak_fc <- featureCounts(files=bamfiles[c(1:4,9:12)],
                               annot.ext = '/disk/user_09/Data/01_TC1/21_M3IN_ca_12h/05_m6A_peak/05_m6A_qc/01_replication/02_peak_merge_group_saf/M3IN_0h_M3IN_12h_peaks.saf',
                               isGTFAnnotationFile = FALSE,
                               useMetaFeatures = FALSE,
                               allowMultiOverlap=FALSE,
                            fracOverlap = 0.5,
                               minMQS = 20, strandSpecific=2,
                               countMultiMappingReads=FALSE,
                               isPairedEnd=TRUE,nthreads=50)

M3IN_0hM3IN_12h_counts <- M3IN_0hM3IN_12h_peak_fc$counts
colnames(M3IN_0hM3IN_12h_counts) <- sample_name[c(1:4,9:12)]

colSums(M3IN_0hM3IN_12h_counts)
M3IN_0hM3IN_12h_cpm <- as.data.frame(t(t(M3IN_0hM3IN_12h_counts)/total_Reads[c(1:4,9:12)]* 1000000))#参考cpm定义
colSums(M3IN_0hM3IN_12h_cpm)

table(rowMeans(M3IN_0hM3IN_12h_cpm)>0)
# FALSE  TRUE 
#   190 19496
M3IN_0hM3IN_12h_cpm <- M3IN_0hM3IN_12h_cpm[rowMeans(M3IN_0hM3IN_12h_cpm)>0,]

M3IN_0hM3IN_12h_cpm_mean <- data.frame(M3IN_0h_input=apply(M3IN_0hM3IN_12h_cpm[,1:2],1,mean),
                             M3IN_0h_ip=apply(M3IN_0hM3IN_12h_cpm[,3:4],1,mean),
                             M3IN_12h_input=apply(M3IN_0hM3IN_12h_cpm[,5:6],1,mean),
                             M3IN_12h_ip=apply(M3IN_0hM3IN_12h_cpm[,7:8],1,mean))

M3IN_0hM3IN_12h_cpm_mean$lfc_M3IN_0h_spikein <- log2((M3IN_0hM3IN_12h_cpm_mean$`M3IN_0h_ip`*scale_facter$m6A_level[1]+1)/(M3IN_0hM3IN_12h_cpm_mean$`M3IN_0h_input`+1))
M3IN_0hM3IN_12h_cpm_mean$lfc_M3IN_12h_spikein <- log2((M3IN_0hM3IN_12h_cpm_mean$`M3IN_12h_ip`*scale_facter$m6A_level[3]+1)/(M3IN_0hM3IN_12h_cpm_mean$`M3IN_12h_input`+1))
summary(M3IN_0hM3IN_12h_cpm_mean$lfc_M3IN_0h_spikein )
summary(M3IN_0hM3IN_12h_cpm_mean$lfc_M3IN_12h_spikein )

M3IN_0hM3IN_12h_cpm_mean$group <- ifelse(M3IN_0hM3IN_12h_cpm_mean$lfc_M3IN_0h_spikein-M3IN_0hM3IN_12h_cpm_mean$lfc_M3IN_12h_spikein>1,"M3IN_0h_high",ifelse(M3IN_0hM3IN_12h_cpm_mean$lfc_M3IN_12h_spikein-M3IN_0hM3IN_12h_cpm_mean$lfc_M3IN_0h_spikein>1,"M3IN_12h_high","NC"))
M3IN_0hM3IN_12h_cpm_mean_filter <- M3IN_0hM3IN_12h_cpm_mean[(M3IN_0hM3IN_12h_cpm_mean$lfc_M3IN_0h_spikein > 1) | (M3IN_0hM3IN_12h_cpm_mean$lfc_M3IN_12h_spikein > 1),]
table(M3IN_0hM3IN_12h_cpm_mean_filter$group)
 # M3IN_0h_high M3IN_12h_high            NC 
 #         7592            97         10571
dir.create("02_peak_m6A_level")
pdf("02_peak_m6A_level/01_m6A_level_change.pdf",width = 8,height = 4)
p1 <- ggplot(M3IN_0hM3IN_4h_cpm_mean_filter, aes(x = lfc_M3IN_4h_spikein, y = lfc_M3IN_0h_spikein,col=group)) +
  geom_point(alpha=0.5,show.legend = F) + # color="#99CC33"
  geom_text(label="9706",inherit.aes = F,x=-0.5,y=4,size=6)+
  geom_text(label="101",inherit.aes = F,x=3.5,y=0.5,size=6)+
  labs(x=expression(m^6*A~level~"in"~M3IN~"4h"), y=expression(m^6*A~level~"in"~M3IN~"0h"))+
  theme_bw()+
  theme(plot.margin = unit(c(0.5,0.5,0.5,0.5), "cm"))+ #调整与图片边缘的距离
  theme(axis.title.x = element_text(size = 20,face="bold"),axis.title.y = element_text(size = 20,face="bold"),plot.title = element_text(size=22,face="bold", hjust = 0.45) ) +
  scale_color_manual(values = c("M3IN_0h_high" = "#eebf99", "M3IN_4h_high" = "#99c7e0"))+
  xlim(-1.7,5)+
  ylim(-1.7,5)+
  coord_fixed(ratio=1)

p2 <- ggplot(M3IN_0hM3IN_12h_cpm_mean_filter, aes(x = lfc_M3IN_12h_spikein, y = lfc_M3IN_0h_spikein,col=group)) +
  geom_point(alpha=0.5,show.legend = F) + # color="#99CC33"
  geom_text(label="97",inherit.aes = F,x=3,y=-0,size=6)+
  geom_text(label="7592",inherit.aes = F,x=0.2,y=3.5,size=6)+
  #labs(x=expression(m^6*A~level~"in"~M3IN~"12h"), y=expression(m^6*A~level~"in"~M3IN~"0h"))+
  labs(x=expression(m^6*A~level~"in"~M3IN~"12h"), y="")+
  theme_bw()+
  theme(plot.margin = unit(c(0.5,0.5,0.5,0.5), "cm"))+ #调整与图片边缘的距离
  theme(axis.title.x = element_text(size = 20,face="bold"),axis.title.y = element_text(size = 20,face="bold"),plot.title = element_text(size=22,face="bold", hjust = 0.45) ) +
  scale_color_manual(values = c("M3IN_0h_high" = "#eebf99", "M3IN_12h_high" = "#99c7e0"))+
  xlim(-1.7,5)+
  ylim(-1.7,5)+
  coord_fixed(ratio=1)
p1+p2
dev.off()

pdf("02_peak_m6A_level/01_m6A_level_change2.pdf",width = 8,height = 4)
p1 <- ggplot(M3IN_0hM3IN_4h_cpm_mean_filter, aes(x = lfc_M3IN_4h_spikein, y = lfc_M3IN_0h_spikein,col=group)) +
  geom_point(alpha=0.5,show.legend = F) + # color="#99CC33"
  geom_text(label="9706",inherit.aes = F,x=-0.5,y=4,size=6)+
  geom_text(label="101",inherit.aes = F,x=3.5,y=0.5,size=6)+
  labs(x=expression(m^6*A~level~"in"~M3IN~"4h"), y=expression(m^6*A~level~"in"~M3IN~"0h"))+
  theme_bw()+
  theme(plot.margin = unit(c(0.5,0.1,0.5,0.5), "cm"))+ #调整与图片边缘的距离
  theme(axis.title.x = element_text(size = 20,face="bold"),axis.title.y = element_text(size = 20,face="bold"),plot.title = element_text(size=22,face="bold", hjust = 0.45) ) +
  scale_color_manual(values = c("M3IN_0h_high" = "#eebf99", "M3IN_4h_high" = "#99c7e0"))+
  xlim(-1.7,5)+
  ylim(-1.7,5)+
  coord_fixed(ratio=1)

p2 <- ggplot(M3IN_0hM3IN_12h_cpm_mean_filter, aes(x = lfc_M3IN_12h_spikein, y = lfc_M3IN_0h_spikein,col=group)) +
  geom_point(alpha=0.5,show.legend = F) + # color="#99CC33"
  geom_text(label="97",inherit.aes = F,x=3,y=-0,size=6)+
  geom_text(label="7592",inherit.aes = F,x=0.2,y=3.5,size=6)+
  #labs(x=expression(m^6*A~level~"in"~M3IN~"12h"), y=expression(m^6*A~level~"in"~M3IN~"0h"))+
  labs(x=expression(m^6*A~level~"in"~M3IN~"12h"), y="")+
  theme_bw()+
  theme(plot.margin = unit(c(0.5,0.5,0.5,0), "cm"))+ #调整与图片边缘的距离
  theme(axis.title.x = element_text(size = 20,face="bold"),axis.title.y = element_text(size = 20,face="bold"),plot.title = element_text(size=22,face="bold", hjust = 0.45) ) +
  scale_color_manual(values = c("M3IN_0h_high" = "#eebf99", "M3IN_12h_high" = "#99c7e0"))+
  xlim(-1.7,5)+
  ylim(-1.7,5)+
  coord_fixed(ratio=1)
p1+p2
dev.off()

```

### homer annotation using basic_pc_lnc.gtf
```{r gtf_basic_pc_lncRNA}
gtf_all <- import("~/reference/annotation/mm39/raw/gencode.vM29.annotation.gtf")
gtf_basic <- import("~/reference/annotation/mm39/raw/gencode.vM29.basic.annotation.gtf")

gtf_all[gtf_all$type=="gene",]
gtf_basic[gtf_basic$type=="gene",]

table(gtf_all[gtf_all$type=="gene",]$gene_type)
gtf_basic[gtf_basic$type=="gene",]

gtf_all[gtf_all$type=="transcript",]
gtf_basic[gtf_basic$type=="transcript",]

table(gtf_all[gtf_all$type=="transcript",]$transcript_type)
table(gtf_basic[gtf_basic$type=="transcript",]$transcript_type)

gtf_basic_pc_lnc <- gtf_basic[gtf_basic$gene_type %in% c("lncRNA","protein_coding"),]

table(gtf_basic_pc_lnc[gtf_basic_pc_lnc$type=="transcript",]$transcript_type)

export(gtf_basic_pc_lnc,"~/reference/annotation/mm39/raw/gencode.vM29.basic.annotation.pc_lnc.gtf","gtf")

```

```{r anno_distribution}
#dir.create("03_annotation")
# read annotation file from macs2 and homer
M3IN_0h_anno <- read.delim('03_annotation/gencode_pc_lnc/M3IN_0h_annotation.txt',header = T)
M3IN_4h_anno <- read.delim('03_annotation/gencode_pc_lnc/M3IN_4h_annotation.txt',header = T)
M3IN_12h_anno <- read.delim('03_annotation/gencode_pc_lnc/M3IN_12h_annotation.txt',header = T)

# save as list
df <- list(M3IN_0h = M3IN_0h_anno$Annotation,M3IN_4h = M3IN_4h_anno$Annotation,M3IN_12h = M3IN_12h_anno$Annotation)

# save separately
lapply(1:3, function(x){
  tmp = sapply(strsplit(df[[x]],split = '\\('),'[',1)
  res = table(tmp)
}) %>% do.call('rbind',.) %>% as.data.frame() -> type_comb

# add name
type_comb$name = c('M3IN_0h','M3IN_4h','M3IN_12h')

# width to length
final <- melt(type_comb)

# factor
final$name <- factor(final$name,levels = c('M3IN_0h','M3IN_4h','M3IN_12h'))


pdf("./03_annotation/02_Distribution_of_m6A_peaks_pc_lnc.pdf")
# draw pie
ggplot(final,aes(x = '',y = value,fill = variable)) +
  geom_col(position = position_fill()) +
  theme_void() +
  theme(legend.position = 'bottom',
        strip.text.x = element_text(size= 20)) +
  facet_wrap(~name) +
  coord_polar(theta = 'y') +
  scale_fill_brewer(palette = 'Set2',name = 'Region types')

# draw bar
ggplot(final,aes(x = name,y = value,fill = variable)) +
  geom_col(position = position_fill()) +
  theme_bw(base_size = 18) +
  scale_y_continuous(labels = scales::percent_format()) +
  theme(legend.position = 'right',
        strip.text.x = element_text(size= 18)) +
  scale_fill_brewer(palette = 'Set2',name = 'Region types') +
  xlab('') + ylab('Percent')

# draw bar
ggplot(final,aes(x = variable,y = value,fill = name)) +
  geom_col(position = position_dodge()) +
  theme_bw(base_size = 18) +
  theme_pubr()+
  theme(legend.position = c(0.8,0.7),
        plot.margin = unit(c(2,0.5,4,0.5),"cm"),
        strip.text.x = element_text(size= 18)) +
  theme(axis.text.x = element_text(angle = 45,vjust = 1,hjust = 1))+
  scale_fill_brewer(palette = 'Set2',name = 'Region types') +
  xlab('') + ylab('number of peaks')
dev.off()
```

```{r radar_plot}
type_comb_perc <- type_comb[,1:5]/rowSums(type_comb[,1:5])
type_comb_perc <- cbind(name=type_comb$name,type_comb_perc)
type_comb_perc

pdf("03_annotation/04_radar_gencode.pdf")
ggradar(type_comb_perc[1:3,],grid.max=0.5,grid.mid = 0.25,
        values.radar = c("","25%","50%"),
        legend.position = "top")
dev.off()
```

### homer mm39 to mm10
```{bash}
mkdir 03_annotation/01_commonpeak_mm10
for sample in {M3IN_0h,M3IN_4h,M3IN_12h}
do
/disk/user_09/software/UCSCtools/liftOver /disk/user_09/Data/01_TC1/21_M3IN_ca_12h/05_m6A_peak/03_bed_merge/03_common_peak_merge/${sample}.bed /disk/user_09/reference/liftover/mm39ToMm10.over.chain.gz 03_annotation/01_commonpeak_mm10/${sample}_common_peaks_mm10.bed 03_annotation/01_commonpeak_mm10/${sample}_common_peaks_mm10_unMapped.bed &
done
```

```{r anno_distribution mm10}

# read annotation file from macs2 and homer
M3IN_0h_anno <- read.delim('03_annotation/mm10/M3IN_0h_annotation.txt',header = T)
M3IN_4h_anno <- read.delim('03_annotation/mm10/M3IN_4h_annotation.txt',header = T)
M3IN_12h_anno <- read.delim('03_annotation/mm10/M3IN_12h_annotation.txt',header = T)

# save as list
df <- list(M3IN_0h = M3IN_0h_anno$Annotation,M3IN_4h = M3IN_4h_anno$Annotation,M3IN_12h = M3IN_12h_anno$Annotation)

# save separately
lapply(1:3, function(x){
  tmp = sapply(strsplit(df[[x]],split = '\\('),'[',1)
  res = table(tmp)
}) %>% do.call('rbind',.) %>% as.data.frame() -> type_comb

# add name
type_comb$name = c('M3IN_0h','M3IN_4h','M3IN_12h')

# width to length
final <- melt(type_comb)

# factor
final$name <- factor(final$name,levels = c('M3IN_0h','M3IN_4h','M3IN_12h'))

pdf("./03_annotation/03_Distribution_of_m6A_peaks_mm10.pdf")
# draw pie
ggplot(final,aes(x = '',y = value,fill = variable)) +
  geom_col(position = position_fill()) +
  theme_void() +
  theme(legend.position = 'bottom',
        strip.text.x = element_text(size= 20)) +
  facet_wrap(~name) +
  coord_polar(theta = 'y') +
  scale_fill_brewer(palette = 'Set2',name = 'Region types')

# draw bar
ggplot(final,aes(x = name,y = value,fill = variable)) +
  geom_col(position = position_fill()) +
  theme_bw(base_size = 18) +
  scale_y_continuous(labels = scales::percent_format()) +
  theme(legend.position = 'right',
        strip.text.x = element_text(size= 18)) +
  scale_fill_brewer(palette = 'Set2',name = 'Region types') +
  xlab('') + ylab('Percent')

# draw bar
ggplot(final,aes(x = variable,y = value,fill = name)) +
  geom_col(position = position_dodge()) +
  theme_bw(base_size = 18) +
  theme_pubr()+
  theme(legend.position = c(0.8,0.7),
        plot.margin = unit(c(2,0.5,4,0.5),"cm"),
        strip.text.x = element_text(size= 18)) +
  theme(axis.text.x = element_text(angle = 45,vjust = 1,hjust = 1))+
  scale_fill_brewer(palette = 'Set2',name = 'Region types') +
  xlab('') + ylab('number of peaks')
dev.off()
```

```{r radar_plot}
type_comb_perc <- type_comb[,1:8]/rowSums(type_comb[,1:8])
type_comb_perc <- cbind(name=type_comb$name,type_comb_perc)
type_comb_perc

pdf("03_annotation/04_radar_mm10.pdf")
ggradar(type_comb_perc[1:2,],grid.max=0.5,grid.mid = 0.25,
        values.radar = c("","25%","50%"),
        legend.position = "top")
dev.off()
```

### metagene
```{r metagene}
common_peak_dir <- "/disk/user_09/Data/01_TC1/21_M3IN_ca_12h/05_m6A_peak/03_bed_merge/03_common_peak_merge/"
M3IN_0h_common_peaks <- import(file.path(common_peak_dir,"M3IN_0h.bed"))
M3IN_4h_common_peaks <- import(file.path(common_peak_dir,"M3IN_4h.bed"))
M3IN_12h_common_peaks <- import(file.path(common_peak_dir,"M3IN_12h.bed"))

# read bed file
stBedFiles <- list(file.path(common_peak_dir,"M3IN_0h.bed"),
                   file.path(common_peak_dir,"M3IN_4h.bed"),
                   file.path(common_peak_dir,"M3IN_12h.bed"))

# prepare annotation file
# txdb <- TxDb.Mmusculus.UCSC.mm10.knownGene

gtffile <- file.path("/disk/user_09/reference/annotation/mm39/raw/gencode.vM29.basic.annotation.pc_lnc.gtf")    #my own gtf
txdb <- makeTxDbFromGFF(gtffile, format="gtf", circ_seqs=character())#read gtf file to construct TxDb

dir.create("04_metagene")

p <- GuitarPlot(txTxdb = txdb,
                   stBedFiles = stBedFiles,
                   headOrtail = F,
                   enableCI = FALSE,
                   mapFilterTranscript = T,
                   pltTxType = c("mrna"),
                   stGroupName = c("M3IN_0h","M3IN_4h","M3IN_12h"))
# draw
pdf("./04_metagene/metagene_profiles_of_m6A_peak.pdf")

p+theme_bw()+
  theme(plot.margin = unit(c(4,0.5,4,0.5),"cm"))
dev.off()



```


### motif
```{r}

pwm.m <- t( read.table(paste0("03_motif_logo/02_motif/M3IN_0h/homerResults/motif23.motif"), header = F, comment.char = ">",col.names = c("A","C","G","U")) )
M3IN_0h_motif <- ggplot() + geom_logo(pwm.m) + theme_logo()+
  ylab("M3IN_0h")+
   theme(axis.text.x = element_blank(),axis.text.y  = element_blank(),axis.title.y = element_text(angle = 0,vjust = 0.5,size=40))

pwm.m <- t( read.table(paste0("03_motif_logo/02_motif/M3IN_4h/homerResults/motif24.motif"), header = F, comment.char = ">",col.names = c("A","C","G","U")) )
M3IN_4h_motif <- ggplot() + geom_logo(pwm.m) + theme_logo()+
  ylab("M3IN_4h")+
   theme(axis.text.x = element_blank(),axis.text.y  = element_blank(),axis.title.y = element_text(angle = 0,vjust = 0.5,size=40))

pwm.m <- t( read.table(paste0("03_motif_logo/01_motif/M3IN_12h/homerResults/motif8.similar1.motif"), header = F, comment.char = ">",col.names = c("A","C","G","U")) )
M3IN_12h_motif <- ggplot() + geom_logo(pwm.m) + theme_logo()+
  ylab("M3IN_12h")+
   theme(axis.text.x = element_blank(),axis.text.y  = element_blank(),axis.title.y = element_text(angle = 0,vjust = 0.5,size=40))

pdf("03_motif_logo/01_motif_logo.pdf",width = 8,height = 9)
M3IN_0h_motif / M3IN_4h_motif / M3IN_12h_motif
dev.off()

```

### peak class 
```{r peak class}
## 0h 4h 
M3IN_0h_4h_peak_bed <- read.table("../04_bed_group_merge/M3IN_0h_M3IN_4h.bed")

M3IN_0h_4h_peak_bed_0h_high <- M3IN_0h_4h_peak_bed[M3IN_0h_4h_peak_bed$V4 %in% rownames(M3IN_0hM3IN_4h_cpm_mean_filter)[M3IN_0hM3IN_4h_cpm_mean_filter$group=="M3IN_0h_high"],]

M3IN_0h_4h_peak_bed_4h_high <- M3IN_0h_4h_peak_bed[M3IN_0h_4h_peak_bed$V4 %in% rownames(M3IN_0hM3IN_4h_cpm_mean_filter)[M3IN_0hM3IN_4h_cpm_mean_filter$group=="M3IN_4h_high"],]

M3IN_0h_4h_peak_bed_nc <- M3IN_0h_4h_peak_bed[M3IN_0h_4h_peak_bed$V4 %in% rownames(M3IN_0hM3IN_4h_cpm_mean_filter)[M3IN_0hM3IN_4h_cpm_mean_filter$group=="NC"],]

tt_wt(M3IN_0h_4h_peak_bed_0h_high,"02_peak_m6A_level/f02_M3IN_0h_4h_peak_0h_high.bed")
tt_wt(M3IN_0h_4h_peak_bed_4h_high,"02_peak_m6A_level/f02_M3IN_0h_4h_peak_4h_high.bed")
tt_wt(M3IN_0h_4h_peak_bed_nc,"02_peak_m6A_level/f02_M3IN_0h_4h_peak_nc.bed")

## 0h 12h 
M3IN_0h_12h_peak_bed <- read.table("../04_bed_group_merge/M3IN_0h_M3IN_12h.bed")

M3IN_0h_12h_peak_bed_0h_high <- M3IN_0h_12h_peak_bed[M3IN_0h_12h_peak_bed$V4 %in% rownames(M3IN_0hM3IN_12h_cpm_mean_filter)[M3IN_0hM3IN_12h_cpm_mean_filter$group=="M3IN_0h_high"],]

M3IN_0h_12h_peak_bed_12h_high <- M3IN_0h_12h_peak_bed[M3IN_0h_12h_peak_bed$V4 %in% rownames(M3IN_0hM3IN_12h_cpm_mean_filter)[M3IN_0hM3IN_12h_cpm_mean_filter$group=="M3IN_12h_high"],]

M3IN_0h_12h_peak_bed_nc <- M3IN_0h_12h_peak_bed[M3IN_0h_12h_peak_bed$V4 %in% rownames(M3IN_0hM3IN_12h_cpm_mean_filter)[M3IN_0hM3IN_12h_cpm_mean_filter$group=="NC"],]

tt_wt(M3IN_0h_12h_peak_bed_0h_high,"02_peak_m6A_level/f02_M3IN_0h_12h_peak_0h_high.bed")
tt_wt(M3IN_0h_12h_peak_bed_12h_high,"02_peak_m6A_level/f02_M3IN_0h_12h_peak_12h_high.bed")
tt_wt(M3IN_0h_12h_peak_bed_nc,"02_peak_m6A_level/f02_M3IN_0h_12h_peak_nc.bed")

```

```{r anno_class_distribution}

# read annotation file from macs2 and homer
M3IN_0h_4h_0h_high_anno <- read.delim('03_annotation/gencode_pc_lnc/M3IN_0h_4h_peak_0h_high_annotation.txt',header = T)
M3IN_0h_4h_4h_high_anno <- read.delim('03_annotation/gencode_pc_lnc/M3IN_0h_4h_peak_4h_high_annotation.txt',header = T)
M3IN_0h_4h_nc_anno <- read.delim('03_annotation/gencode_pc_lnc/M3IN_0h_4h_peak_nc_annotation.txt',header = T)

# save as list
df <- list(M3IN_0h_4h_0h_high = M3IN_0h_4h_0h_high_anno$Annotation,
           M3IN_0h_4h_4h_high = M3IN_0h_4h_4h_high_anno$Annotation,
           M3IN_0h_4h_nc = M3IN_0h_4h_nc_anno$Annotation)

# save separately
lapply(1:3, function(x){
  tmp = sapply(strsplit(df[[x]],split = '\\('),'[',1)
  res = table(tmp)
}) %>% do.call('rbind',.) %>% as.data.frame() -> type_comb

# add name
type_comb$name = c('0h_high','4h_high','nc')

# width to length
final <- melt(type_comb)

# factor
final$name <- factor(final$name,levels = c('0h_high','4h_high','nc'))


pdf("./03_annotation/05_Distribution_of_m6A_peaks_pc_lnc_M3IN_0h_4h_class.pdf")
# draw pie
ggplot(final,aes(x = '',y = value,fill = variable)) +
  geom_col(position = position_fill()) +
  theme_void() +
  theme(legend.position = 'bottom',
        strip.text.x = element_text(size= 20)) +
  facet_wrap(~name) +
  coord_polar(theta = 'y') +
  scale_fill_brewer(palette = 'Set2',name = 'Region types')

# draw bar
ggplot(final,aes(x = name,y = value,fill = variable)) +
  geom_col(position = position_fill()) +
  theme_bw(base_size = 18) +
  scale_y_continuous(labels = scales::percent_format()) +
  theme(legend.position = 'right',
        strip.text.x = element_text(size= 18)) +
  scale_fill_brewer(palette = 'Set2',name = 'Region types') +
  xlab('') + ylab('Percent')

# draw bar
ggplot(final,aes(x = variable,y = value,fill = name)) +
  geom_col(position = position_dodge()) +
  theme_bw(base_size = 18) +
  theme_pubr()+
  theme(legend.position = c(0.8,0.7),
        plot.margin = unit(c(2,0.5,4,0.5),"cm"),
        strip.text.x = element_text(size= 18)) +
  theme(axis.text.x = element_text(angle = 45,vjust = 1,hjust = 1))+
  scale_fill_brewer(palette = 'Set2',name = 'Region types') +
  xlab('') + ylab('number of peaks')
dev.off()

## 12h
# read annotation file from macs2 and homer
M3IN_0h_12h_0h_high_anno <- read.delim('03_annotation/gencode_pc_lnc/M3IN_0h_12h_peak_0h_high_annotation.txt',header = T)
M3IN_0h_12h_12h_high_anno <- read.delim('03_annotation/gencode_pc_lnc/M3IN_0h_12h_peak_12h_high_annotation.txt',header = T)
M3IN_0h_12h_nc_anno <- read.delim('03_annotation/gencode_pc_lnc/M3IN_0h_12h_peak_nc_annotation.txt',header = T)

# save as list
df <- list(M3IN_0h_12h_0h_high = M3IN_0h_12h_0h_high_anno$Annotation,
           M3IN_0h_12h_12h_high = M3IN_0h_12h_12h_high_anno$Annotation,
           M3IN_0h_12h_nc = M3IN_0h_12h_nc_anno$Annotation)

# save separately
lapply(1:3, function(x){
  tmp = sapply(strsplit(df[[x]],split = '\\('),'[',1)
  res = table(tmp)
}) %>% do.call('rbind',.) %>% as.data.frame() -> type_comb

# add name
type_comb$name = c('0h_high','12h_high','nc')

# width to length
final <- melt(type_comb)

# factor
final$name <- factor(final$name,levels = c('0h_high','12h_high','nc'))


pdf("./03_annotation/05_Distribution_of_m6A_peaks_pc_lnc_M3IN_0h_12h_class.pdf")
# draw pie
ggplot(final,aes(x = '',y = value,fill = variable)) +
  geom_col(position = position_fill()) +
  theme_void() +
  theme(legend.position = 'bottom',
        strip.text.x = element_text(size= 20)) +
  facet_wrap(~name) +
  coord_polar(theta = 'y') +
  scale_fill_brewer(palette = 'Set2',name = 'Region types')

# draw bar
ggplot(final,aes(x = name,y = value,fill = variable)) +
  geom_col(position = position_fill()) +
  theme_bw(base_size = 18) +
  scale_y_continuous(labels = scales::percent_format()) +
  theme(legend.position = 'right',
        strip.text.x = element_text(size= 18)) +
  scale_fill_brewer(palette = 'Set2',name = 'Region types') +
  xlab('') + ylab('Percent')

# draw bar
ggplot(final,aes(x = variable,y = value,fill = name)) +
  geom_col(position = position_dodge()) +
  theme_bw(base_size = 18) +
  theme_pubr()+
  theme(legend.position = c(0.8,0.7),
        plot.margin = unit(c(2,0.5,4,0.5),"cm"),
        strip.text.x = element_text(size= 18)) +
  theme(axis.text.x = element_text(angle = 45,vjust = 1,hjust = 1))+
  scale_fill_brewer(palette = 'Set2',name = 'Region types') +
  xlab('') + ylab('number of peaks')
dev.off()
```

```{r exon anno GO}
M3IN_0h_12h_0h_high_anno$annotation <- sapply(strsplit(M3IN_0h_12h_0h_high_anno$Annotation,split = " "),"[",1)

M3IN_0h_12h_0h_high_anno_ETP <- M3IN_0h_12h_0h_high_anno[M3IN_0h_12h_0h_high_anno$annotation %in% c("exon","TTS","promoter-TSS"),]

M3IN_0h_12h_0h_high_ETP_genes <- M3IN_0h_12h_0h_high_anno_ETP$Gene.Name[!duplicated(M3IN_0h_12h_0h_high_anno_ETP$Gene.Name)]

M3IN_0h_12h_0h_high_ETP_pc_genes <- M3IN_0h_12h_0h_high_anno_ETP$Gene.Name[!duplicated(M3IN_0h_12h_0h_high_anno_ETP$Gene.Name) & M3IN_0h_12h_0h_high_anno_ETP$Gene.Type=="protein_coding"]

dir.create("03_annotation/02_M3IN_0h_12h_class")
tt_wt(M3IN_0h_12h_0h_high_ETP_pc_genes,"03_annotation/02_M3IN_0h_12h_class/M3IN_0h_12h_0h_high_ETP_pc_genes.txt")
```

```{r}
length(intersect(M3IN_0h_12h_0h_high_generegion_genes,read.table("~/Data/01_TC1/04_TC1_mRNA/03_TPgene/T_gene2.txt")$V1))
length(intersect(M3IN_0h_12h_0h_high_generegion_genes,read.table("~/Data/01_TC1/04_TC1_mRNA/03_TPgene/P_gene2.txt")$V1))



```

### QNB peak

```{r QNB}
# When size.factor is not NA
total_number_reads_0h_ip <- c(74606690,65858934)
total_number_reads_4h_ip <- c(63084688,75080218)
total_number_reads_12h_ip <- c(63731668,62110020)
total_number_reads_0h_input <- c(65411554,59680388)
total_number_reads_4h_input <- c(55155274,55244282)
total_number_reads_12h_input <- c(55453230,61066238)

# calculate the number of reads for a "standard" library
standard_library_size <- exp(mean(log( c(total_number_reads_0h_ip,
                                  total_number_reads_4h_ip,
                                  total_number_reads_12h_ip,
                                  total_number_reads_0h_input,
                                  total_number_reads_4h_input,
                                  total_number_reads_12h_input))))

nf_spikein <- c(0.697398363,0.752975137,1.255850546,1.157329973,1.074906923,1.061539057)
```

## M3IN 0h 4h
```{r gene featureCounts cpm}
dir.create("05_m6A_peak_QNB")
bam_dir <- "/disk/user_09/Data/01_TC1/21_M3IN_ca_12h/03_hisat2_mapping/03_bam_merge/"
sample_name <- paste(rep(c("M3IN_0h","M3IN_4h","M3IN_12h"),each=4),
                  rep(c("input","ip"),each=2,times=3),
                  rep(c("rep1","rep2"),times=6),
                 sep="_")

bamfiles <- paste0(bam_dir,sample_name,".bam")

total_Reads <- read.table("/disk/user_09/Data/01_TC1/21_M3IN_ca_12h/03_hisat2_mapping/03_bam_merge/flagstat/flagstat_num_summary.txt")$V1/2

total_Reads <- total_Reads[c(1,2,7,8,3,4,9,10,5,6,11,12)]

## M3IN_0h_4h
M3IN_0h_4h_fc <- featureCounts(files=bamfiles[1:8],
                               annot.ext = '/disk/user_09/Data/01_TC1/21_M3IN_ca_12h/05_m6A_peak/04_bed_group_merge/M3IN_0h_M3IN_4h.saf',
                               isGTFAnnotationFile = F,
                               allowMultiOverlap=FALSE,
                            minOverlap = 30,
                               minMQS = 20, strandSpecific=2,
                               countMultiMappingReads=FALSE,
                            fraction = FALSE,
                               isPairedEnd=TRUE,nthreads=50)

M3IN_0h_4h_counts <- M3IN_0h_4h_fc$counts[,1:8]
colnames(M3IN_0h_4h_counts) <- sample_name[1:8]

tt_wt(M3IN_0h_4h_counts,"05_m6A_peak_QNB/M3IN_0h_4h_counts.txt",row.names = T,col.names = T)
colSums(M3IN_0h_4h_counts)

table(rowSums(M3IN_0h_4h_counts>10)>=2)
# FALSE  TRUE 
#     7 20380
#M3IN_0h_4h_counts_filter <- M3IN_0h_4h_counts[rowSums(M3IN_0h_4h_counts>10)>=2,]
M3IN_0h_4h_counts_filter <- M3IN_0h_4h_counts

```

```{r M3IN 0h 4h}
M3IN_0h_ip_0h_4h_counts = M3IN_0h_4h_counts[,3:4]
M3IN_0h_input_0h_4h_counts = M3IN_0h_4h_counts[,1:2]
M3IN_4h_ip_0h_4h_counts = M3IN_0h_4h_counts[,7:8]
M3IN_4h_input_0h_4h_counts = M3IN_0h_4h_counts[,5:6]

head(M3IN_0h_ip_0h_4h_counts)

# calculate the sample size factor based on the total number of reads
size.factor.spikein <- list(control_ip=total_number_reads_0h_ip*nf_spikein[1:2]/standard_library_size,
                    treated_ip=total_number_reads_4h_ip*nf_spikein[3:4]/standard_library_size,
                   control_input=total_number_reads_0h_input/standard_library_size,
                   treated_input=total_number_reads_4h_input/standard_library_size)

dir.create("05_m6A_peak_QNB/M3IN_0h_4h")
result_0h_4h = qnbtest(M3IN_0h_ip_0h_4h_counts, M3IN_4h_ip_0h_4h_counts, M3IN_0h_input_0h_4h_counts, M3IN_4h_input_0h_4h_counts, mode="per-condition",output.dir = "05_m6A_peak_QNB/M3IN_0h_4h",size.factor = size.factor.spikein)
tt_wt(result_0h_4h,"05_m6A_peak_QNB/M3IN_0h_4h/result_0h_4h.txt",row.names = T,col.names = T)

```

## M3IN_0h_12h
```{r gene featureCounts cpm}

M3IN_0h_12h_fc <- featureCounts(files=bamfiles[c(1:4,9:12)],
                               annot.ext = '/disk/user_09/Data/01_TC1/21_M3IN_ca_12h/05_m6A_peak/04_bed_group_merge/M3IN_0h_M3IN_12h.saf',
                               isGTFAnnotationFile = F,
                               allowMultiOverlap=FALSE,
                            minOverlap = 30,
                               minMQS = 20, strandSpecific=2,
                               countMultiMappingReads=FALSE,
                            fraction = FALSE,
                               isPairedEnd=TRUE,nthreads=50)

M3IN_0h_12h_counts <- M3IN_0h_12h_fc$counts
colnames(M3IN_0h_12h_counts) <- sample_name[c(1:4,9:12)]

tt_wt(M3IN_0h_12h_counts,"05_m6A_peak_QNB/M3IN_0h_12h_counts.txt",row.names = T,col.names = T)
colSums(M3IN_0h_12h_counts)

table(rowSums(M3IN_0h_12h_counts>10)>=2)
# FALSE  TRUE 
#     7 20380
#M3IN_0h_12h_counts_filter <- M3IN_0h_12h_counts[rowSums(M3IN_0h_12h_counts>10)>=2,]
M3IN_0h_12h_counts_filter <- M3IN_0h_12h_counts

```

```{r M3IN 0h 12h}

M3IN_0h_ip_0h_12h_counts = M3IN_0h_12h_counts[,3:4]
M3IN_0h_input_0h_12h_counts = M3IN_0h_12h_counts[,1:2]
M3IN_12h_ip_0h_12h_counts = M3IN_0h_12h_counts[,7:8]
M3IN_12h_input_0h_12h_counts = M3IN_0h_12h_counts[,5:6]

# calculate the sample size factor based on the total number of reads
size.factor.spikein <- list(control_ip=total_number_reads_0h_ip*nf_spikein[1:2]/standard_library_size,
                    treated_ip=total_number_reads_12h_ip*nf_spikein[5:6]/standard_library_size,
                   control_input=total_number_reads_0h_input/standard_library_size,
                   treated_input=total_number_reads_12h_input/standard_library_size)

dir.create("05_m6A_peak_QNB/M3IN_0h_12h")
result_0h_12h = qnbtest(M3IN_0h_ip_0h_12h_counts, M3IN_12h_ip_0h_12h_counts, M3IN_0h_input_0h_12h_counts, M3IN_12h_input_0h_12h_counts, mode="per-condition",output.dir = "05_m6A_peak_QNB/M3IN_0h_12h",size.factor = size.factor.spikein)

tt_wt(result_0h_12h,"05_m6A_peak_QNB/M3IN_0h_12h/result_0h_12h.txt",row.names = T,col.names = T)

```












