---
title: "r01_DEG"
author: "Tang Li"
date: '2022-10-16'
output: html_document
---

```{r setup, include=FALSE}
# ~/Data/01_TC1/21_M3IN_ca_12h/03_hisat2_mapping/04_DEG/04_DEG.Rproj
knitr::opts_chunk$set(echo = TRUE)
library(DESeq2)
library(GEOquery)
library(GenomicFeatures)
library(BiocParallel)
library(GenomicAlignments)
library(rtracklayer)
library(NMF)
library(parallel)
library(Rsubread)
#BiocManager::install("RUVSeq")
library(RUVSeq)
library(stringr)
library(data.table)
library(ggplot2)
library(ggprism)
library(RColorBrewer)
library(ttFunctions)
library("ggrepel")
library(ggthemes)
library(ggsci)
library(ggpubr)
library(ComplexHeatmap)

require("knitr")
opts_knit$set(root.dir = "~/Data/01_TC1/22_M3IN_mRNA_12h/03_hisat2_mapping/04_DEG/")

```

## featurecount

```{r}
opts_knit$set(root.dir = "~/Data/01_TC1/22_M3IN_mRNA_12h/03_hisat2_mapping/04_DEG/")

#设置工作路径
dir1='/disk/user_09/Data/01_TC1/22_M3IN_mRNA_12h/03_hisat2_mapping/03_bam_merge/'
sample_name1 <- paste(rep(c("M3IN"),6),
                      rep(c("0h","4h","12h"),each=2),
                      rep(c("rep1","rep2"),times=3),
                      sep = "_")

fls <- file.path(dir1,paste0(sample_name1,'.bam'))
file.exists(fls)

fc_exon <- tt_featureCounts_gtf(fls = fls,annot.ext = "~/reference/annotation/mm39/raw/gencode.vM29.annotation.gtf")

fc_exon_count <- as.data.frame(fc_exon$counts)
colnames(fc_exon_count) <- sample_name1
rownames(fc_exon_count) <- fc_exon$annotation$GeneID

fc_exon_annotation <- fc_exon$annotation
# 
tt_wt(fc_exon_count,file.path("express_matrix_raw.txt")) 
tt_wt(fc_exon_annotation,file.path("annotation_raw.txt")) 
```

```{r gene filter & get expression matrix}
fc_exon_count_pc <- fc_exon_count[fc_exon_annotation$gene_type=="protein_coding",]

#get annotation
mm_anno <- rtracklayer::import('/disk/user_09/reference/annotation/mm39/raw/gencode.vM29.annotation_ERCC.gtf')
mm_anno <- rtracklayer::as.data.frame(mm_anno)

# Non-Overlapping Exon Length ----

# then for each gene, reduce all the exons to a set of non overlapping exons, 
# calculate their lengths (widths) and sum then
txdb <- makeTxDbFromGFF(file = '/disk/user_09/reference/annotation/mm39/raw/gencode.vM29.annotation.gtf')

exonByGene <- exonsBy(txdb,by='gene')
exonic.gene.sizes <- sum(width(reduce(exonByGene)))
head(exonic.gene.sizes)

#modify annotation
mm_gene <- mm_anno[mm_anno["type"] == "gene",]
exonic.gene.sizes <- exonic.gene.sizes[mm_gene$gene_id]
mm_gene$length <- exonic.gene.sizes
rownames(mm_gene) <- mm_gene$gene_id

table(rownames(fc_exon_count_pc) %in% rownames(mm_gene))
table(rownames(mm_gene) %in% rownames(fc_exon_count_pc))
mm_gene <- mm_gene[rownames(fc_exon_count_pc),]

table(duplicated(mm_gene$gene_id))
table(duplicated(mm_gene$gene_name))
mm_gene <- mm_gene[!duplicated(mm_gene$gene_name),]

fc_exon_count_pc <- fc_exon_count_pc[rownames(mm_gene),]
table(rownames(fc_exon_count_pc) %in% rownames(mm_gene))

rownames(fc_exon_count_pc) <- mm_gene$gene_name
rownames(mm_gene) <- mm_gene$gene_name

set_matrix <- fc_exon_count_pc

```

### DEG for input samples

```{r betweenLaneNormalization}
set_matrix_DEG <- set_matrix[,1:6]

# 对表达量进行筛选
filter <- rowSums(set_matrix_DEG>10)>=2
table(filter)
set_matrix_DEG <- set_matrix_DEG[filter,]

## 批次校正
# batch <- c(rep("p6si",6),rep("p16si",10),rep("mRNA",8),rep("XEN",6))
# library(sva)
# express_sva <- ComBat_seq(counts=as.matrix(set_matrix),batch=batch)
# set_matrix <- express_sva
###
x <- as.factor(rep(c("M3IN_0h","M3IN_4h","M3IN_12h"),each=2))
set <- newSeqExpressionSet(as.matrix(set_matrix_DEG),
                           phenoData = data.frame(x, row.names=colnames(set_matrix_DEG)))
set


colors <- brewer.pal(18, "Paired")
plotRLE(set, outline=FALSE, ylim=c(-2, 2), col=colors[x])
plotPCA(set, col=colors[x], cex=1.2)

set <- betweenLaneNormalization(set, which="upper")
plotRLE(set, outline=FALSE, ylim=c(-4, 4), col=colors[x])
plotPCA(set, col=colors[x], cex=1.2)

express_matrix <- set@assayData$normalizedCounts

colnames(express_matrix)
tt_wt(express_matrix,"express_matrix.txt",row.names=T,col.names = T)
```

```{r get tpm_matrix}
#convert to TPM

table(rownames(mm_gene) %in% rownames(express_matrix))
mm_gene_tpm <- mm_gene[rownames(express_matrix),]
tpm_matrix <- do.call(cbind,base::lapply(colnames(express_matrix),FUN = function(x){
  counts <- express_matrix[,x]
  effLen <- mm_gene_tpm$length
  rate <- log(counts) - log(effLen)
  denom <- log(sum(exp(rate)))
  return(exp(rate-denom+log(1e6)))
}))

rownames(tpm_matrix) <- rownames(express_matrix)
colnames(tpm_matrix) <- colnames(express_matrix)
table(colSums(tpm_matrix))

write.table(tpm_matrix,file="tpm_matrix_input.txt",quote = F,sep = "\t")
```

## DESeq2
```{r create dds object and plot pca}

dir.create("01_pca")

#load meta data
sample_name <- colnames(express_matrix)
meta_data <- data.frame(rep = c(rep(c("rep1","rep2"),times=3)),
                        stage = factor(rep(c("M3IN_0h","M3IN_4h","M3IN_12h"),each=2),
                                       levels = c("M3IN_0h","M3IN_4h","M3IN_12h")),
                        row.names = sample_name,stringsAsFactors = TRUE)
meta_data <- as.matrix(meta_data)
meta_data 

dds <- DESeqDataSetFromMatrix(countData = express_matrix,colData = meta_data,design = ~ stage)

#pca plot

rlog <- rlog(dds,blind = FALSE)

pcaData <- plotPCA(rlog,intgroup = c("stage"),returnData = TRUE,ntop=500)
pcaData
percentVar <- round(100 * attr(pcaData,"percentVar"))
percentVar

p1 <- ggplot(pcaData,aes(x = PC1, y = PC2, color=stage)) +
  geom_point(size=3) +
  geom_label(aes(label=stage,vjust=1))+
  xlab(paste0("PC1:",percentVar[1],"% variance")) +
  ylab(paste0("PC2:",percentVar[2],"% variance")) +
  xlim(-6,8)+
  ylim(-4,4)+
  coord_fixed(ratio=percentVar[2]/percentVar[1]) +
  ggtitle("PCA plot mRNA input")+
  theme_bw()+
  scale_color_prism(palette = "floral")
  #scale_color_manual(values = c("M3IN_0h" = "#facf87", "M3IN_4h" = "#d1d1ed", "M3IN_12h" = "#b9dbf4"))


pdf("01_pca/01_pca_rlog.pdf",width = 8,height = 7)
print(p1)
dev.off()
```

```{r DEG analysis}
dir.create("02_DEG_result")
dir.create("03_DEG_heatmap")
dir.create("04_DEG_volcano")

#DEG
dds <- DESeq(dds)
resultsNames(dds)
```


```{r DEG result filter1}
#filter1
#### 


sn <- c("M3IN_0h","M3IN_4h","M3IN_12h")

res_M3IN_0h_4h <- tt_DEG_result(dds,"M3IN_0h","M3IN_4h")
res_M3IN_0h_12h <- tt_DEG_result(dds,"M3IN_0h","M3IN_12h")
res_M3IN_4h_12h <- tt_DEG_result(dds,"M3IN_4h","M3IN_12h")

res_M3IN_0h_4h_filter_1 <- tt_DEG_filter(res_M3IN_0h_4h,i = "M3IN_0h",j="M3IN_4h",lfcThreshold = 1)
res_M3IN_0h_12h_filter_1 <- tt_DEG_filter(res_M3IN_0h_12h,i = "M3IN_0h",j="M3IN_12h",lfcThreshold = 1)
res_M3IN_4h_12h_filter_1 <- tt_DEG_filter(res_M3IN_4h_12h,i = "M3IN_4h",j="M3IN_12h",lfcThreshold = 1)
  
tt_DEG_heatmap(res_M3IN_0h_4h_filter_1,"M3IN_0h","M3IN_4h",tpm_matrix,meta_data,lfcThreshold = 1,pdf_width = 5)
tt_DEG_heatmap(res_M3IN_0h_12h_filter_1,"M3IN_0h","M3IN_12h",tpm_matrix,meta_data,lfcThreshold = 1)
tt_DEG_heatmap(res_M3IN_4h_12h_filter_1,"M3IN_4h","M3IN_12h",tpm_matrix,meta_data,lfcThreshold = 1)

tt_DEG_volcano(res_M3IN_0h_4h,"0h to","4h",lfcThreshold =1,pdf_width = 8,pdf_height = 6)
tt_DEG_volcano(res_M3IN_0h_12h,"0h to","12h",lfcThreshold =1,pdf_width = 8,pdf_height = 6)
tt_DEG_volcano(res_M3IN_4h_12h,"4h to","12h",lfcThreshold =1,pdf_width = 8,pdf_height = 6)
```

```{r transKEGG}
transKEGG <- function(i){
  df <- bitr( rownames(get(paste0("tpm_newOrder",i))), 
              fromType = "SYMBOL", toType = c( "ENTREZID"), OrgDb = org.Mm.eg.db )
  KEGG=enrichKEGG(gene=df$ENTREZID, organism = 'mmu', keyType = 'kegg', pvalueCutoff = 0.05,pAdjustMethod = 'BH', 
                  minGSSize = 10,maxGSSize = 500,qvalueCutoff = 0.1,use_internal_data = T)
  if (nrow(as.data.frame(KEGG))==0){
    return(KEGG)
  }
  write.csv(summary(KEGG),paste0("05_diff_M3IN/km_4/01_GO/", i,"_KEGGenrich_significant.csv"),row.names =F)
  
  KEGG_bar <- barplot(KEGG,title=paste0("KEGG_",i),drop = TRUE, showCategory =10)
  KEGG_dot <- dotplot(KEGG,title=paste0("KEGG_",i),showCategory = 10) 
  pdf(file=paste0("05_diff_M3IN/km_4/01_GO/KEGG_bar_",i,".pdf"), width=7, height=8)
  print(KEGG_bar)
  print(KEGG_dot)
  dev.off()
  return(KEGG)
}
transGO <- function(i){
  df <- bitr( rownames(get(paste0("tpm_newOrder",i))), 
              fromType = "SYMBOL", toType = c( "ENTREZID"), OrgDb = org.Mm.eg.db )
  ego <- enrichGO(
    gene = df$ENTREZID,
    keyType = "ENTREZID",
    OrgDb = org.Mm.eg.db,
    ont = "all", # "BP", "MF", "CC"
    pAdjustMethod = "BH",
    pvalueCutoff = 0.05,
    qvalueCutoff = 0.05,
    readable = TRUE
  )
  
  if (nrow(as.data.frame(ego))==0){
    return(ego)
  }
  
  # significant
  ego_df <- as.data.frame(ego)
  dat <- ego_df[ego_df$p.adjust < 0.05, ]
  dat <- dat[order(dat$p.adjust, decreasing = F), ]
  all_GO=dat
  write.csv(all_GO, paste0("05_diff_M3IN/km_4/01_GO/",i,"_GOenrich_significant.csv"), row.names = F)
  
  go_bar <- barplot(ego,title=paste0("GO_",i),drop = TRUE, showCategory =10,split="ONTOLOGY")+ 
    facet_grid(ONTOLOGY~., scale='free')
  go_dot <- dotplot(ego, orderBy = "x", showCategory =10,split="ONTOLOGY",
                    title = paste0("GO_",i)) + facet_grid(ONTOLOGY~., scale='free')
  
  pdf(file=paste0("05_diff_M3IN/km_4/01_GO/Go_bar_",i,".pdf"), width=9, height=8)
  print(go_bar)
  print(go_dot)
  dev.off()
  return(all_GO)
}
```

```{r diff_M3IN cluster}
dir.create("05_diff_M3IN")
M3IN_diff_geneset_filter1 <- c(rownames(res_M3IN_0h_4h_filter_1),
                       rownames(res_M3IN_4h_12h_filter_1),
                       rownames(res_M3IN_0h_12h_filter_1))

M3IN_diff_geneset_filter1 <- M3IN_diff_geneset_filter1[!duplicated(M3IN_diff_geneset_filter1)]

tpm_M3IN_diff <- tpm_matrix[M3IN_diff_geneset_filter1,]


set.seed(1000)
tpm_M3IN_diff_mean <- data.frame(M3IN_0h=apply(tpm_M3IN_diff[,1:2],1,mean),
                                 M3IN_4h=apply(tpm_M3IN_diff[,3:4],1,mean),
                                 M3IN_12h=apply(tpm_M3IN_diff[,5:6],1,mean))
########### row_km=6
dir.create("05_diff_M3IN/km_4")

heatmap_M3IN_diff <- Heatmap(matrix = t(scale(t(tpm_M3IN_diff_mean))),name="z-score",
                                show_column_names = TRUE,show_row_names=FALSE,
                                cluster_rows = TRUE,cluster_columns = F,
                                width = unit(4,'inches'),
                                row_km = 4,
                                col=circlize::colorRamp2(c(-1.5,0,1.5), 
                                                         colorRampPalette(c("#4875aa", "white", "#d7604c"))(3)))


pdf(paste0("05_diff_M3IN/km_4/heatmap_km4.pdf"))
heatmap_M3IN_diff <- draw(heatmap_M3IN_diff)
dev.off()


tpm_newOrder1 <- tpm_M3IN_diff_mean[row_order(heatmap_M3IN_diff)$`1`,]
tpm_newOrder2 <- tpm_M3IN_diff_mean[row_order(heatmap_M3IN_diff)$`2`,]
tpm_newOrder3 <- tpm_M3IN_diff_mean[row_order(heatmap_M3IN_diff)$`3`,]
tpm_newOrder4 <- tpm_M3IN_diff_mean[row_order(heatmap_M3IN_diff)$`4`,]
#tpm_newOrder5 <- tpm_M3IN_diff_mean[row_order(heatmap_M3IN_diff)$`5`,]
#tpm_newOrder6 <- tpm_M3IN_diff_mean[row_order(heatmap_M3IN_diff)$`6`,]
#tpm_newOrder7 <- tpm_M3IN_diff_mean[row_order(heatmap_M3IN_diff2)$`7`,]
#tpm_newOrder8 <- tpm_M3IN_diff_mean[row_order(heatmap_M3IN_diff2)$`8`,]
#tpm_newOrder9 <- tpm_M3IN_diff_mean[row_order(heatmap_M3IN_diff2)$`9`,]

gene_anno <- HeatmapAnnotation(cluster = c(rep("2",dim(tpm_newOrder2)[1]),
                                           rep("1",dim(tpm_newOrder1)[1]),
                                           rep("4",dim(tpm_newOrder4)[1]),
                                           rep("3",dim(tpm_newOrder3)[1])),which = 'row')
colnames(tpm_M3IN_diff_mean)
heatmap_M3IN_diff_order <- Heatmap(matrix = t(scale(t(tpm_M3IN_diff_mean[c(rownames(tpm_newOrder2),
                                            rownames(tpm_newOrder1),
                                            rownames(tpm_newOrder4),
                                            rownames(tpm_newOrder3)),]))),
        name="z-score",
                                show_column_names = TRUE,show_row_names=FALSE,
                                cluster_rows = F,cluster_columns = F,
        left_annotation = gene_anno,
                                width = unit(3,'inches'),
                                col=circlize::colorRamp2(c(-1.5,0,1.5), 
                                                         colorRampPalette(c("#4875aa", "white", "#d7604c"))(3)))


pdf("05_diff_M3IN/km_4/heatmap_km4_order.pdf")
heatmap_M3IN_diff_order <- draw(heatmap_M3IN_diff_order)
dev.off()


write.table(tpm_newOrder1,"05_diff_M3IN/km_4/tpm_newOrder1.txt",quote = F)
write.table(tpm_newOrder2,"05_diff_M3IN/km_4/tpm_newOrder2.txt",quote = F)
write.table(tpm_newOrder3,"05_diff_M3IN/km_4/tpm_newOrder3.txt",quote = F)
write.table(tpm_newOrder4,"05_diff_M3IN/km_4/tpm_newOrder4.txt",quote = F)

write.table(rownames(tpm_newOrder1),"05_diff_M3IN/km_4/newOrder1_genes.txt",quote = F,row.names = F,col.names = F)
write.table(rownames(tpm_newOrder2),"05_diff_M3IN/km_4/newOrder2_genes.txt",quote = F,row.names = F,col.names = F)
write.table(rownames(tpm_newOrder3),"05_diff_M3IN/km_4/newOrder3_genes.txt",quote = F,row.names = F,col.names = F)
write.table(rownames(tpm_newOrder4),"05_diff_M3IN/km_4/newOrder4_genes.txt",quote = F,row.names = F,col.names = F)

options(stringsAsFactors = FALSE)
library(ggrepel)
library(stringr)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(ggplot2)
library(clusterProfiler)
library(org.Mm.eg.db)
keytypes(org.Mm.eg.db)

library(data.table)

# 1. id transform
# biomart
library(biomaRt)
library(data.table)
# bitr in clusterProfiler

dir.create("05_diff_M3IN/km_4/01_GO")



for (i in 1:4){
   assign(paste0("Go_",i),transGO(i))
}

for (i in 1:4){
        assign(paste0("KEGG_",i),transKEGG(i))
}

```
```{r gsea}


```

```{r diff_M3IN cluster}
dir.create("05_diff_M3IN")
M3IN_diff_geneset_filter1 <- c(rownames(res_M3IN_0h_4h_filter_1),
                       rownames(res_M3IN_4h_12h_filter_1),
                       rownames(res_M3IN_0h_12h_filter_1))

M3IN_diff_geneset_filter1 <- M3IN_diff_geneset_filter1[!duplicated(M3IN_diff_geneset_filter1)]

tpm_M3IN_diff <- tpm_matrix[M3IN_diff_geneset_filter1,]


set.seed(1000)
tpm_M3IN_diff_mean <- data.frame(M3IN_0h=apply(tpm_M3IN_diff[,1:2],1,mean),
                                 M3IN_4h=apply(tpm_M3IN_diff[,3:4],1,mean),
                                 M3IN_12h=apply(tpm_M3IN_diff[,5:6],1,mean))
########### row_km=6
dir.create("05_diff_M3IN/km_4")

heatmap_M3IN_diff <- Heatmap(matrix = t(scale(t(tpm_M3IN_diff_mean))),name="z-score",
                                show_column_names = TRUE,show_row_names=FALSE,
                                cluster_rows = TRUE,cluster_columns = F,
                                width = unit(4,'inches'),
                                row_km = 6,
                                col=circlize::colorRamp2(c(-1.5,0,1.5), 
                                                         colorRampPalette(c("#4875aa", "white", "#d7604c"))(3)))


pdf(paste0("05_diff_M3IN/km_4/heatmap_km6.pdf"))
heatmap_M3IN_diff <- draw(heatmap_M3IN_diff)
dev.off()


tpm_newOrder1 <- tpm_M3IN_diff_mean[row_order(heatmap_M3IN_diff)$`1`,]
tpm_newOrder2 <- tpm_M3IN_diff_mean[row_order(heatmap_M3IN_diff)$`2`,]
tpm_newOrder3 <- tpm_M3IN_diff_mean[row_order(heatmap_M3IN_diff)$`3`,]
tpm_newOrder4 <- tpm_M3IN_diff_mean[row_order(heatmap_M3IN_diff)$`4`,]
tpm_newOrder5 <- tpm_M3IN_diff_mean[row_order(heatmap_M3IN_diff)$`5`,]
tpm_newOrder6 <- tpm_M3IN_diff_mean[row_order(heatmap_M3IN_diff)$`6`,]
#tpm_newOrder7 <- tpm_M3IN_diff_mean[row_order(heatmap_M3IN_diff2)$`7`,]
#tpm_newOrder8 <- tpm_M3IN_diff_mean[row_order(heatmap_M3IN_diff2)$`8`,]
#tpm_newOrder9 <- tpm_M3IN_diff_mean[row_order(heatmap_M3IN_diff2)$`9`,]

gene_anno <- HeatmapAnnotation(cluster = c(rep("2",dim(tpm_newOrder2)[1]),
                                           rep("1",dim(tpm_newOrder1)[1]),
                                           rep("3",dim(tpm_newOrder3)[1]),
                                           rep("6",dim(tpm_newOrder6)[1]),
                                           rep("5",dim(tpm_newOrder5)[1]),
                                           rep("4",dim(tpm_newOrder4)[1])),which = 'row')
colnames(tpm_M3IN_diff_mean)
heatmap_M3IN_diff_order <- Heatmap(matrix = t(scale(t(tpm_M3IN_diff_mean[c(rownames(tpm_newOrder2),
                                            rownames(tpm_newOrder1),
                                            rownames(tpm_newOrder3),
                                            rownames(tpm_newOrder6),
                                            rownames(tpm_newOrder5),
                                            rownames(tpm_newOrder4)),]))),
        name="z-score",
                                show_column_names = TRUE,show_row_names=FALSE,
                                cluster_rows = F,cluster_columns = F,
        left_annotation = gene_anno,
                                width = unit(3,'inches'),
                                col=circlize::colorRamp2(c(-1.5,0,1.5), 
                                                         colorRampPalette(c("#4875aa", "white", "#d7604c"))(3)))


pdf("05_diff_M3IN/km_4/heatmap_km6_order.pdf")
heatmap_M3IN_diff_order <- draw(heatmap_M3IN_diff_order)
dev.off()


write.table(tpm_newOrder1,"05_diff_M3IN/km_4/tpm_newOrder1.txt",quote = F)
write.table(tpm_newOrder2,"05_diff_M3IN/km_4/tpm_newOrder2.txt",quote = F)
write.table(tpm_newOrder3,"05_diff_M3IN/km_4/tpm_newOrder3.txt",quote = F)
write.table(tpm_newOrder4,"05_diff_M3IN/km_4/tpm_newOrder4.txt",quote = F)
write.table(tpm_newOrder5,"05_diff_M3IN/km_4/tpm_newOrder5.txt",quote = F)
write.table(tpm_newOrder6,"05_diff_M3IN/km_4/tpm_newOrder6.txt",quote = F)

write.table(rownames(tpm_newOrder1),"05_diff_M3IN/km_4/newOrder1_genes.txt",quote = F,row.names = F,col.names = F)
write.table(rownames(tpm_newOrder2),"05_diff_M3IN/km_4/newOrder2_genes.txt",quote = F,row.names = F,col.names = F)
write.table(rownames(tpm_newOrder3),"05_diff_M3IN/km_4/newOrder3_genes.txt",quote = F,row.names = F,col.names = F)
write.table(rownames(tpm_newOrder4),"05_diff_M3IN/km_4/newOrder4_genes.txt",quote = F,row.names = F,col.names = F)
write.table(rownames(tpm_newOrder5),"05_diff_M3IN/km_4/newOrder5_genes.txt",quote = F,row.names = F,col.names = F)
write.table(rownames(tpm_newOrder6),"05_diff_M3IN/km_4/newOrder6_genes.txt",quote = F,row.names = F,col.names = F)

options(stringsAsFactors = FALSE)
library(ggrepel)
library(stringr)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(ggplot2)
library(clusterProfiler)
library(org.Mm.eg.db)
keytypes(org.Mm.eg.db)

library(data.table)

# 1. id transform
# biomart
library(biomaRt)
library(data.table)
# bitr in clusterProfiler

dir.create("05_diff_M3IN/km_4/01_GO")



for (i in 1:6){
   assign(paste0("Go_",i),transGO(i))
}




for (i in 1:6){
        assign(paste0("KEGG_",i),transKEGG(i))
}
```

```{r}
library(ComplexHeatmap)
Heatmap(matrix = t(scale(t(tpm_matrix[rownames(tpm_matrix) %in% read.table("~/Data/01_TC1/00_gene_set/Toti_genes")$V1,]))),name="z-score",
        show_column_names = TRUE,show_row_names=T,
        cluster_rows = T,cluster_columns = F,
        width = unit(4,'inches'),
        col=circlize::colorRamp2(c(-1.5,0,1.5), 
                                 colorRampPalette(c("#4875aa", "white", "#d7604c"))(3)))

Heatmap(matrix = t(scale(t(tpm_matrix[rownames(tpm_matrix) %in% read.table("~/Data/01_TC1/00_gene_set/Pluri_genes")$V1,]))),name="z-score",
        show_column_names = TRUE,show_row_names=T,
        cluster_rows = T,cluster_columns = F,
        width = unit(4,'inches'),
        col=circlize::colorRamp2(c(-1.5,0,1.5), 
                                 colorRampPalette(c("#4875aa", "white", "#d7604c"))(3)))

Heatmap(matrix = t(scale(t(tpm_matrix[rownames(tpm_matrix) %in% read.table("~/Data/01_TC1/00_gene_set/PrE_genes")$V1,]))),name="z-score",
        show_column_names = TRUE,show_row_names=T,
        cluster_rows = T,cluster_columns = F,
        width = unit(4,'inches'),
        col=circlize::colorRamp2(c(-1.5,0,1.5), 
                                 colorRampPalette(c("#4875aa", "white", "#d7604c"))(3)))

Heatmap(matrix = t(scale(t(tpm_matrix[rownames(tpm_matrix) %in% read.table("~/Data/01_TC1/00_gene_set/Trophectoderm_markers")$V1,]))),name="z-score",
        show_column_names = TRUE,show_row_names=T,
        cluster_rows = T,cluster_columns = F,
        width = unit(4,'inches'),
        col=circlize::colorRamp2(c(-1.5,0,1.5), 
                                 colorRampPalette(c("#4875aa", "white", "#d7604c"))(3)))


```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
